{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "code_folding": [],
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Load packages & model\n",
    "import cobra\n",
    "import xlrd\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from libsbml import *\n",
    "from pprint import pprint\n",
    "from cobra.flux_analysis import gapfill\n",
    "model=cobra.io.read_sbml_model(\"models/iPAU_final.xml\") #import the version of GENRE from Alina's work to manipulate\n",
    "from cobra import *\n",
    "from cobra.flux_analysis import *\n",
    "from LJD_Functions_DDP import *\n",
    "from DDP_Functions  import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial Growth:  15.729799767441344\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/core/group.py:110 \u001b[1;31mUserWarning\u001b[0m: need to pass in a list\n"
     ]
    }
   ],
   "source": [
    "#DDP changes made to update to iPau21 - this cell corresponds to DDP sheet of iPau21 Change Log\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#_________________________________________________________________________________________\n",
    "#Changes based on DDP research\n",
    "\n",
    "#Remove reactions rJB00275, rJB00280, rJB00281, EX_cpd00147_c because they were leftover demand reactions and not carrying flux\n",
    "model.reactions.rJB00275.remove_from_model()\n",
    "model.reactions.rJB00280.remove_from_model()\n",
    "model.reactions.rJB00281.remove_from_model() #line 53 of LJD\n",
    "model.reactions.EX_cpd00147_c.remove_from_model()\n",
    "\n",
    "#Change cJB0034_e to ModelSeed ID (cpd00212_e)\n",
    "model.metabolites.cJB00034_e.id = 'cpd00212_e' #line 3 DDP - corresponds to line 10 and 11 of LJD\n",
    "model.metabolites.cpd00212_e.notes['NOTES'] = '*DDP 2/17/20 - replaced cJB00034 with seed cpd00212'\n",
    "model.metabolites.cpd00212_e.annotation['seed.compound'] = 'cpd00212'\n",
    "model.metabolites.cpd00212_e.annotation['kegg.compound'] = 'C00247'\n",
    "model.metabolites.cpd00212_e.annotation['AraCyc'] = 'CPD-9569'\n",
    "model.metabolites.cpd00212_e.annotation['bigg.metabolite'] = 'srb__L'\n",
    "model.metabolites.cpd00212_e.annotation['BrachyCyc'] = 'CPD-9569'\n",
    "model.metabolites.cpd00212_e.annotation['MetaCyc'] = 'CPD-9569'\n",
    "\n",
    "#Change cPY00161_c to ModelSeed ID (cpd23005_c)\n",
    "model.metabolites.cPY00161_c.id = 'cpd23005_c' #line 4 DDP\n",
    "model.metabolites.cpd23005_c.notes['NOTES'] = '*DDP 2/17/20 - replaced cPY00161 with seed cpd23005'\n",
    "model.metabolites.cpd23005_c.annotation['MetaCyc'] = 'CPD-11571'\n",
    "\n",
    "#Rename reactions with ModelSEED ID's found\n",
    "model.reactions.rJB00259.id = 'rxn36698' #line 2\n",
    "model.reactions.rxn36698.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00259 with seed rxn36698. Same reaction *PY 06/11/14 Changed rightarrow to leqrightarrow *JAB 04/16/14 - ub8gequb9 fix; ASB 03/27/14 - tentative subsystem assignment; JAB 01/28/14 - MetaCyc indicates irreversibility'\n",
    "model.reactions.rxn36698.annotation['BranchCyc'] = 'DIHYDROOROTATE-DEHYDROGENASE-RXN'\n",
    "model.reactions.rxn36698.annotation['ChlamyCyc'] = 'DIHYDROOROTATE-DEHYDROGENASE-RXN'\n",
    "model.reactions.rxn36698.annotation['MetaCyc'] = 'DIHYDROOROTATE-DEHYDROGENASE-RXN; RXN0-6491'\n",
    "model.reactions.rxn36698.annotation['PlantCyc'] = 'DIHYDROOROTATE-DEHYDROGENASE-RXN; RXN0-6491'\n",
    "model.reactions.rxn36698.annotation['seed.reaction'] = 'rxn36698'\n",
    "model.reactions.rxn36698.annotation['ec-code'] = '1.3.5.2'\n",
    "\n",
    "model.reactions.rJB00260.id = 'rxn37440' #line 3\n",
    "model.reactions.rxn37440.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00260 with seed rxn37440. Same reaction *PY 06/11/14 Changed rightarrow to rightarrow *JAB 04/16/14 - ub8gequb9 fix, changed reaction direction based on MetaCyc'\n",
    "model.reactions.rxn37440.annotation['BrachyCyc'] = 'RXN-9927; RXN0-5258'\n",
    "model.reactions.rxn37440.annotation['ChlamyCyc'] = 'RXN0-5258'\n",
    "model.reactions.rxn37440.annotation['MetaCyc'] = 'RXN0-5258; RXN0-5260'\n",
    "model.reactions.rxn37440.annotation['PlantCyc'] = 'RXN-9927; RXN0-5258; RXN0-5260'\n",
    "model.reactions.rxn37440.annotation['seed.reaction'] = 'rxn37440'\n",
    "\n",
    "model.reactions.rJB00262.id = 'rxn36761' #line 4\n",
    "model.reactions.rxn36761.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00262 with seed rxn36761. Same reaction *PY 06/11/14 Changed rightarrow to rightarrow *JAB 04/16/14 - ub8gequb9 fix; ASB 03/25/14 - tentative subsystem assignment; JAB 01/31/14 - switched reaction directionality according to MetaCyc'\n",
    "model.reactions.rxn36761.annotation['MetaCyc'] = 'MALATE-DEHYDROGENASE-ACCEPTOR-RXN'\n",
    "model.reactions.rxn36761.annotation['PlantCyc'] = 'MALATE-DEHYDROGENASE-ACCEPTOR-RXN'\n",
    "model.reactions.rxn36761.annotation['seed.reaction'] = 'rxn36761'\n",
    "model.reactions.rxn36761.annotation['ec-code'] = '1.1.5.4'\n",
    "\n",
    "model.reactions.rJB00265.id = 'rxn36840' #line 5\n",
    "model.reactions.rxn36840.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00265 with seed rxn36840. Same reaction. Updated GPR to include PA14_48440 and PA14_51990 associated with EC. *PY 06/11/14 Changed rightarrow to rightarrow *JAB 04/16/14 - ub8gequb9 fix; ASB 03/25/14 - tentative subsystem assignment; JAB 01/31/14 - switched reaction directionality to match MetaCyc'\n",
    "model.reactions.rxn36840.annotation['MetaCyc'] = 'NQOR-RXN; RXN0-271'\n",
    "model.reactions.rxn36840.annotation['ec-code'] = '1.6.5.2'\n",
    "model.reactions.rxn36840.annotation['seed.reaction'] = 'rxn36840'\n",
    "model.reactions.rxn36840.gene_reaction_rule = 'PA14_48440 or PA14_51990 or PA14_65760'\n",
    "\n",
    "model.reactions.rJB00266.id = 'rxn37076' #line 6\n",
    "model.reactions.rxn37076.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00266 with seed rxn37076. Same reaction *PY 06/11/14 Changed rightarrow to rightarrow *JAB 04/16/14 - ub8gequb9 fix; JAB 01/28/14 - bounds were incorrect (forcing opposite, incorrect irreversibility)'\n",
    "model.reactions.rxn37076.annotation['MetaCyc'] = 'RXN-11496'\n",
    "model.reactions.rxn37076.annotation['seed.reaction'] = 'rxn37076'\n",
    "model.reactions.rxn37076.annotation['ec-code'] = '1.2.5.1'\n",
    "model.reactions.rxn37076.gene_reaction_rule = \"(PA14_37340 or PA14_69925)\"\n",
    "\n",
    "model.reactions.rJB00276.id = 'rxn10902' #line 7\n",
    "model.reactions.rxn10902.notes['NOTES'] = '*DDP 2/17/20 - replaced rJB00276 with seed rxn10902. Same reaction *JAB 06/29/14 - zwf shows much higher cofactor binding specificity for NADP, but higher enzyme activity when NAD is bound; Fuhrer and Sauer show NAD cofactor activity in P. fluorescens and theorize that dual isozyme specificities may help correct redox balance - P. fluorescens and P. aeruginosa genes are nearly identical (not sure which might prefer NAD, but is suspected)'\n",
    "model.reactions.rxn10902.references = {'REFERENCES': 'Ma, J., Hager, P., Howell, M., Phibbs, P., and Hassett, D. (1998). Cloning and characterization of the Pseudomonas aeruginosa zwf gene encoding glucose-6-phosphate dehydrogenase, an enzyme important in resistance to methyl viologen (Paraquat). Journal of Bacteriology, 180(7), 1741–1749. Retrieved from http://jb.asm.org/content/180/7/1741.short; Fuhrer, T., and Sauer, U. (2009). Different biochemical mechanisms ensure network-wide balancing of reducing equivalents in microbial metabolism. Journal of Bacteriology, 191(7), 2112–21. doi:10.1128/JB.01523-08'}\n",
    "model.reactions.rxn10902.annotation['kegg.compound'] = 'R10520'\n",
    "model.reactions.rxn10902.annotation['MetaCyc'] = 'RXN-16652'\n",
    "model.reactions.rxn10902.annotation['iIT341'] = 'G6PDH1'\n",
    "model.reactions.rxn10902.annotation['ec-code'] = '1.1.1.363'\n",
    "model.reactions.rxn10902.annotation['seed.reaction'] = 'rxn10902'\n",
    "\n",
    "model.reactions.rPY00216.id = 'rxn42682' #line 8\n",
    "model.reactions.rxn42682.notes['NOTES'] = '*DDP 2/17/20 - replaced rPY00216 with seed rxn42682. Same reaction *PY 04/17/14 Added for pyoverdine I synthesis (virulence factor). Implemented from Metacyc (https://bit.ly/2pd5GJV). Gene is also known as pvdA.'\n",
    "model.reactions.rxn42682.annotation['MetaCyc'] = 'RXN-11128'\n",
    "model.reactions.rxn42682.annotation['kegg.reaction'] = 'R10789'\n",
    "model.reactions.rxn42682.annotation['ec-code'] = '1.14.13.195'\n",
    "model.reactions.rxn42682.annotation['seed.reaction'] = 'rxn42682'\n",
    "\n",
    "model.reactions.rPY00220.id = 'rxn21407' #line 9\n",
    "model.reactions.rxn21407.notes['NOTES'] = '*DDP 2/17/20 - replaced rPY00220 with seed rxn21407. Same reaction *PY 04/17/14 Added for pyochelin synthesis (virulence factor). Implemented from Metacyc (https://bit.ly/2pdtwWb). Genes are also called pchD, pchE, pchF, and pchG, respectively. Please see the Metacyc link for what is currently known about pyochelin synthesis. Compound name for pyochelin taken from SEED master list.'\n",
    "model.reactions.rxn21407.annotation['MetaCyc'] = 'RXN-11114'\n",
    "model.reactions.rxn21407.annotation['ec-code'] = '6.2.1.61'\n",
    "model.reactions.rxn21407.annotation['seed.reaction'] = 'rxn21407'\n",
    "\n",
    "#Add exchange reactions for extracellular metabolites that don't have them\n",
    "reaction6 = Reaction('EX_cJB00127_e') #line 3\n",
    "reaction6.name = 'EX cis-2-Decenoic acid'\n",
    "reaction6.subsystem = 'Exchange'\n",
    "reaction6.lower_bound = 0\n",
    "reaction6.upper_bound = 0\n",
    "reaction6.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction6.add_metabolites({model.metabolites.cJB00127_e: -1.0})\n",
    "reaction6.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction6])\n",
    "\n",
    "\n",
    "reaction7 = Reaction('EX_cpd00047_e') #line 4\n",
    "reaction7.name = 'EX Formate'\n",
    "reaction7.subsystem = 'Exchange'\n",
    "reaction7.lower_bound = 0\n",
    "reaction7.upper_bound = 0\n",
    "reaction7.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction7.add_metabolites({model.metabolites.cpd00047_e: -1.0})\n",
    "reaction7.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction7])\n",
    "\n",
    "reaction8 = Reaction('EX_cpd00059_e') #line 5\n",
    "reaction8.name = 'EX L-Ascorbate'\n",
    "reaction8.subsystem = 'Exchange'\n",
    "reaction8.lower_bound = 0\n",
    "reaction8.upper_bound = 0\n",
    "reaction8.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction8.add_metabolites({model.metabolites.cpd00059_e: -1.0})\n",
    "reaction8.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction8])\n",
    "\n",
    "reaction9 = Reaction('EX_cpd00085_e') #line 6\n",
    "reaction9.name = 'EX beta-Alanine'\n",
    "reaction9.subsystem = 'Exchange'\n",
    "reaction9.lower_bound = 0\n",
    "reaction9.upper_bound = 0\n",
    "reaction9.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction9.add_metabolites({model.metabolites.cpd00085_e: -1.0})\n",
    "reaction9.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction9])\n",
    "\n",
    "reaction10 = Reaction('EX_cpd00136_e') #line 7\n",
    "reaction10.name = 'EX 4-Hydroxybenzoate'\n",
    "reaction10.subsystem = 'Exchange'\n",
    "reaction10.lower_bound = 0\n",
    "reaction10.upper_bound = 0\n",
    "reaction10.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction10.add_metabolites({model.metabolites.cpd00136_e: -1.0})\n",
    "reaction10.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction10])\n",
    "\n",
    "reaction11 = Reaction('EX_cpd00147_e') #line 8\n",
    "reaction11.name = 'EX 5-Methylthioadenosine'\n",
    "reaction11.subsystem = 'Exchange'\n",
    "reaction11.lower_bound = 0\n",
    "reaction11.upper_bound = 0\n",
    "reaction11.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction11.add_metabolites({model.metabolites.cpd00147_e: -1.0})\n",
    "reaction11.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction11])\n",
    "\n",
    "reaction12 = Reaction('EX_cpd00152_e') #line 9\n",
    "reaction12.name = 'EX Agmatine'\n",
    "reaction12.subsystem = 'Exchange'\n",
    "reaction12.lower_bound = 0\n",
    "reaction12.upper_bound = 0\n",
    "reaction12.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction12.add_metabolites({model.metabolites.cpd00152_e: -1.0})\n",
    "reaction12.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction12])\n",
    "\n",
    "reaction13 = Reaction('EX_cpd00170_e') #line 10\n",
    "reaction13.name = 'EX Gluconolactone'\n",
    "reaction13.subsystem = 'Exchange'\n",
    "reaction13.lower_bound = 0\n",
    "reaction13.upper_bound = 0\n",
    "reaction13.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction13.add_metabolites({model.metabolites.cpd00170_e: -1.0})\n",
    "reaction13.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction13])\n",
    "\n",
    "reaction14 = Reaction('EX_cpd00186_e') #line 11\n",
    "reaction14.name = 'EX D-Glutamate'\n",
    "reaction14.subsystem = 'Exchange'\n",
    "reaction14.lower_bound = 0\n",
    "reaction14.upper_bound = 0\n",
    "reaction14.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction14.add_metabolites({model.metabolites.cpd00186_e: -1.0})\n",
    "reaction14.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction14])\n",
    "\n",
    "reaction15 = Reaction('EX_cpd00214_e') #line 12\n",
    "reaction15.name = 'EX Palmitate'\n",
    "reaction15.subsystem = 'Exchange'\n",
    "reaction15.lower_bound = 0\n",
    "reaction15.upper_bound = 0\n",
    "reaction15.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction15.add_metabolites({model.metabolites.cpd00214_e: -1.0})\n",
    "reaction15.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction15])\n",
    "\n",
    "reaction16 = Reaction('EX_cpd00222_e') #line 13\n",
    "reaction16.name = 'EX GLCN'\n",
    "reaction16.subsystem = 'Exchange'\n",
    "reaction16.lower_bound = 0\n",
    "reaction16.upper_bound = 0\n",
    "reaction16.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction16.add_metabolites({model.metabolites.cpd00222_e: -1.0})\n",
    "reaction16.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction16])\n",
    "\n",
    "reaction17 = Reaction('EX_cpd00229_e') #line 14\n",
    "reaction17.name = 'EX Glycolaldehyde'\n",
    "reaction17.subsystem = 'Exchange'\n",
    "reaction17.lower_bound = 0\n",
    "reaction17.upper_bound = 0\n",
    "reaction17.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction17.add_metabolites({model.metabolites.cpd00229_e: -1.0})\n",
    "reaction17.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction17])\n",
    "\n",
    "reaction18 = Reaction('EX_cpd00247_e') #line 15\n",
    "reaction18.name = 'EX Orotate'\n",
    "reaction18.subsystem = 'Exchange'\n",
    "reaction18.lower_bound = 0\n",
    "reaction18.upper_bound = 0\n",
    "reaction18.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction18.add_metabolites({model.metabolites.cpd00247_e: -1.0})\n",
    "reaction18.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction18])\n",
    "\n",
    "reaction19 = Reaction('EX_cpd00266_e') #line 16\n",
    "reaction19.name = 'EX Carnitine'\n",
    "reaction19.subsystem = 'Exchange'\n",
    "reaction19.lower_bound = 0\n",
    "reaction19.upper_bound = 0\n",
    "reaction19.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction19.add_metabolites({model.metabolites.cpd00266_e: -1.0})\n",
    "reaction19.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction19])\n",
    "\n",
    "reaction20 = Reaction('EX_cpd00363_e') #line 17\n",
    "reaction20.name = 'EX Ethanol'\n",
    "reaction20.subsystem = 'Exchange'\n",
    "reaction20.lower_bound = 0\n",
    "reaction20.upper_bound = 0\n",
    "reaction20.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction20.add_metabolites({model.metabolites.cpd00363_e: -1.0})\n",
    "reaction20.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction20])\n",
    "\n",
    "reaction21 = Reaction('EX_cpd00379_e') #line 18\n",
    "reaction21.name = 'EX Glutarate'\n",
    "reaction21.subsystem = 'Exchange'\n",
    "reaction21.lower_bound = 0\n",
    "reaction21.upper_bound = 0\n",
    "reaction21.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction21.add_metabolites({model.metabolites.cpd00379_e: -1.0})\n",
    "reaction21.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction21])\n",
    "\n",
    "reaction22 = Reaction('EX_cpd00480_e') #line 19 / LJD line 54\n",
    "reaction22.name = 'EX 2-Dehydro-D-gluconate'\n",
    "reaction22.subsystem = 'Exchange'\n",
    "reaction22.lower_bound = 0\n",
    "reaction22.upper_bound = 0\n",
    "reaction22.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction22.add_metabolites({model.metabolites.cpd00480_e: -1.0})\n",
    "reaction22.notes['NOTES'] = '*LJD 12/12/18 - replaced secretion only rJB00281 with an exchange. Setting the upper bound to 1000 and lower bound to -10 on minimal media does not result in model growth or 2-deyhdro-d-gluconate secretion. JAB 07/09/14 - added permanently to model to allow gluconate/2-ketogluconate accumulation (turn reactions on in appropriate simulations, forward only)'\n",
    "model.add_reactions([reaction22])\n",
    "\n",
    "reaction23 = Reaction('EX_cpd00666_e') #line 20\n",
    "reaction23.name = 'EX Tartrate'\n",
    "reaction23.subsystem = 'Exchange'\n",
    "reaction23.lower_bound = 0\n",
    "reaction23.upper_bound = 0\n",
    "reaction23.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction23.add_metabolites({model.metabolites.cpd00666_e: -1.0})\n",
    "reaction23.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction23])\n",
    "\n",
    "reaction24 = Reaction('EX_cpd00794_e') #line 21\n",
    "reaction24.name = 'EX TRHL'\n",
    "reaction24.subsystem = 'Exchange'\n",
    "reaction24.lower_bound = 0\n",
    "reaction24.upper_bound = 0\n",
    "reaction24.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction24.add_metabolites({model.metabolites.cpd00794_e: -1.0,})\n",
    "reaction24.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction24])\n",
    "\n",
    "reaction25 = Reaction('EX_cpd00870_e') #line 22\n",
    "reaction25.name = 'EX gamma-butyrobetaine'\n",
    "reaction25.subsystem = 'Exchange'\n",
    "reaction25.lower_bound = 0\n",
    "reaction25.upper_bound = 0\n",
    "reaction25.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction25.add_metabolites({model.metabolites.cpd00870_e: -1.0,})\n",
    "reaction25.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction25])\n",
    "\n",
    "reaction26 = Reaction('EX_cpd01293_e') #line 23\n",
    "reaction26.name = 'EX 5-Oxoproline'\n",
    "reaction26.subsystem = 'Exchange'\n",
    "reaction26.lower_bound = 0\n",
    "reaction26.upper_bound = 0\n",
    "reaction26.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction26.add_metabolites({model.metabolites.cpd01293_e: -1.0,})\n",
    "reaction26.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction26])\n",
    "\n",
    "reaction27 = Reaction('EX_cpd01502_e') #line 24\n",
    "reaction27.name = 'EX Citraconate'\n",
    "reaction27.subsystem = 'Exchange'\n",
    "reaction27.lower_bound = 0\n",
    "reaction27.upper_bound = 0\n",
    "reaction27.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction27.add_metabolites({model.metabolites.cpd01502_e: -1.0,})\n",
    "reaction27.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction27])\n",
    "\n",
    "reaction28 = Reaction('EX_cpd01630_e') #line 25\n",
    "reaction28.name = 'EX cis,cis-Muconate'\n",
    "reaction28.subsystem = 'Exchange'\n",
    "reaction28.lower_bound = 0\n",
    "reaction28.upper_bound = 0\n",
    "reaction28.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction28.add_metabolites({model.metabolites.cpd01630_e: -1.0,})\n",
    "reaction28.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction28])\n",
    "\n",
    "reaction29 = Reaction('EX_cpd01947_e') #line 26\n",
    "reaction29.name = 'EX BDOH'\n",
    "reaction29.subsystem = 'Exchange'\n",
    "reaction29.lower_bound = 0\n",
    "reaction29.upper_bound = 0\n",
    "reaction29.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction29.add_metabolites({model.metabolites.cpd01947_e: -1.0,})\n",
    "reaction29.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction29])\n",
    "\n",
    "reaction30 = Reaction('EX_cpd02012_e') #line 27\n",
    "reaction30.name = 'EX N-Formyl-L-methionine'\n",
    "reaction30.subsystem = 'Exchange'\n",
    "reaction30.lower_bound = 0\n",
    "reaction30.upper_bound = 0\n",
    "reaction30.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction30.add_metabolites({model.metabolites.cpd02012_e: -1.0,})\n",
    "reaction30.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction30])\n",
    "\n",
    "reaction31 = Reaction('EX_cpd03294_e') #line 28\n",
    "reaction31.name = 'EX Aerobactin'\n",
    "reaction31.subsystem = 'Exchange'\n",
    "reaction31.lower_bound = 0\n",
    "reaction31.upper_bound = 0\n",
    "reaction31.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction31.add_metabolites({model.metabolites.cpd03294_e: -1.0,})\n",
    "reaction31.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction31])\n",
    "\n",
    "reaction32 = Reaction('EX_cpd03726_e') #line 29\n",
    "reaction32.name = 'EX Fe-enterochlin'\n",
    "reaction32.subsystem = 'Exchange'\n",
    "reaction32.lower_bound = 0\n",
    "reaction32.upper_bound = 0\n",
    "reaction32.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction32.add_metabolites({model.metabolites.cpd03726_e: -1.0,})\n",
    "reaction32.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction32])\n",
    "\n",
    "reaction33 = Reaction('EX_cpd03847_e') #line 30\n",
    "reaction33.name = 'EX Myristic acid'\n",
    "reaction33.subsystem = 'Exchange'\n",
    "reaction33.lower_bound = 0\n",
    "reaction33.upper_bound = 0\n",
    "reaction33.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction33.add_metabolites({model.metabolites.cpd03847_e: -1.0,})\n",
    "reaction33.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction33])\n",
    "\n",
    "reaction34 = Reaction('EX_cpd04099_e') #line 31\n",
    "reaction34.name = 'EX Phosphonate'\n",
    "reaction34.subsystem = 'Exchange'\n",
    "reaction34.lower_bound = 0\n",
    "reaction34.upper_bound = 0\n",
    "reaction34.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction34.add_metabolites({model.metabolites.cpd04099_e: -1.0,})\n",
    "reaction34.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction34])\n",
    "\n",
    "reaction35 = Reaction('EX_cpd07061_e') #line 32\n",
    "reaction35.name = 'EX Stachydrine'\n",
    "reaction35.subsystem = 'Exchange'\n",
    "reaction35.lower_bound = 0\n",
    "reaction35.upper_bound = 0\n",
    "reaction35.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction35.add_metabolites({model.metabolites.cpd07061_e: -1.0,})\n",
    "reaction35.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction35])\n",
    "\n",
    "reaction36 = Reaction('EX_cpd11608_e') #line 33\n",
    "reaction36.name = 'EX Peptide'\n",
    "reaction36.subsystem = 'Exchange'\n",
    "reaction36.lower_bound = 0\n",
    "reaction36.upper_bound = 0\n",
    "reaction36.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction36.add_metabolites({model.metabolites.cpd11608_e: -1.0,})\n",
    "reaction36.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction36])\n",
    "\n",
    "reaction37 = Reaction('EX_cpd15901_e') #line 34\n",
    "reaction37.name = 'EX 2-oxohexanedioic acid'\n",
    "reaction37.subsystem = 'Exchange'\n",
    "reaction37.lower_bound = 0\n",
    "reaction37.upper_bound = 0\n",
    "reaction37.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction37.add_metabolites({model.metabolites.cpd15901_e: -1.0,})\n",
    "reaction37.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction37])\n",
    "\n",
    "reaction38 = Reaction('EX_cpd17074_e') #line 35\n",
    "reaction38.name = 'EX Acetylated Alginate'\n",
    "reaction38.subsystem = 'Exchange'\n",
    "reaction38.lower_bound = 0\n",
    "reaction38.upper_bound = 0\n",
    "reaction38.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction38.add_metabolites({model.metabolites.cpd17074_e: -1.0,})\n",
    "reaction38.notes['NOTES'] = '*DDP 2/14/20 - Added exchange reaction that was missing'\n",
    "model.add_reactions([reaction38])\n",
    "#_________________________________________________________________________________________\n",
    "\n",
    "#Make changes based on LJD spreadsheet (changesForOnlineRepo_ForDawson)\n",
    "\n",
    "#Remove metabolites\n",
    "model.metabolites.cPY00016_e.remove_from_model() #line 19\n",
    "model.metabolites.cPY00116_e.remove_from_model() #line 20\n",
    "        \n",
    "#Add cpd03587_e\n",
    "met2 = Metabolite('cpd03587_e',\n",
    "                 formula = 'C110H196N2O39P2',\n",
    "                 name = 'kdo2-lipid a',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -6)\n",
    "model.add_metabolites([met2])\n",
    "model.metabolites.cpd03587_e.annotation = {'kegg.compound':'C00139','sbo': 'SBO:0000247'} #line 21\n",
    "\n",
    "#Change metabolite formulas/charges\n",
    "model.metabolites.cpd16660_c.charge = 1 #line 7\n",
    "model.metabolites.cpd16660_c.formula = 'C4H8NO2' #line7\n",
    "\n",
    "model.metabolites.cpd11595_c.formula = 'H2CrO4' #line 8\n",
    "\n",
    "model.metabolites.cpd11595_e.formula = 'H2CrO4' #line 9\n",
    "\n",
    "model.metabolites.cpd17016_c.charge = -2 #line 12\n",
    "model.metabolites.cpd17016_c.annotation['kegg.compound'] = 'C00139' #line 12\n",
    "model.metabolites.cpd17016_c.annotation['mSim'] = 'cellb' #line 12\n",
    "\n",
    "model.metabolites.cpd17017_c.charge = 0 #line 13\n",
    "model.metabolites.cpd17017_c.annotation['kegg.compound'] = 'C00138' #line 13\n",
    "model.metabolites.cpd17017_c.annotation['mSim'] = 'itacon' #line 13\n",
    "\n",
    "#Rename Reactions - LJD recorded these as remove one and add another\n",
    "model.reactions.rJB00226.id = 'rxn09171' #line18/19\n",
    "model.reactions.rxn09171.notes['NOTES'] = '*LJD 12/12/18 - replaced rJB00226 with rxn09171 from model seed (identical). JAB 12/07/14 - added PA14 duplication for PA3234; putP (PA0783) mainly proline transport but propionate transport is also suggested, MctC of C. glutamicum matches acetate permease PA3234 (Jolkver et al), sodium symport reaction already formulated for Burk but not in SEED DB'\n",
    "model.reactions.rxn09171.references = {'REFERENCES':'Jolkver, E., Emer, D., Ballan, S., Krämer, R., Eikmanns, B. J., & Marin, K. (2009). Identification and characterization of a bacterial transport system for the uptake of pyruvate, propionate, and acetate in Corynebacterium glutamicum. Journal of Bacteriology, 191(3), 940–8. doi:10.1128/JB.01155-08'}\n",
    "\n",
    "model.reactions.rJB00234.id = 'rxn10559' #line21/22\n",
    "model.reactions.rxn10559.name = '4-Hydroxyphenylacetate transport'\n",
    "model.reactions.rxn10559.notes['NOTES'] = '*LJD 12/12/18 - converted to seed rxn10559. Seed says it is periplasmic, but I have it extracellular to intracellular. JAB 01/13/14 - added from Burk'\n",
    "model.reactions.rxn10559.annotation['bigg.metabolite'] = '4HPACt2r; 4HPHACtex; 4HPHACtpp; HPACt2r'\n",
    "model.reactions.rxn10559.annotation['iAF692'] = 'HPACt2r'\n",
    "model.reactions.rxn10559.annotation['iMA945'] = '4HPHACtex; 4HPHACtpp'\n",
    "model.reactions.rxn10559.annotation['iRR1083'] = '4HPACt2r'\n",
    "\n",
    "model.reactions.rJB00235.id = 'rxn08262' #line23/24\n",
    "model.reactions.rxn08262.notes['NOTES'] = '*LJD 12/12/18 - replaced rJB000235 with seed rxn08262. Same reaction. JAB 01/13/14 - added from Burk'\n",
    "\n",
    "model.reactions.rJB00279.id = 'rxn07967'#line 51/52\n",
    "model.reactions.rxn07967.name ='2-dehydro-D-gluconate reductase (NADH)'\n",
    "model.reactions.rxn07967.annotation['bigg.metabolite'] ='2DGLCNRx'\n",
    "model.reactions.rxn07967.annotation['ec-code'] ='1.1.1.215'\n",
    "model.reactions.rxn07967.notes['NOTES'] = '*LJD 12/12/18 - replaced rJB00279 with rxn07967 from seed. Same reaction slightly different name. JAB 07/09/14 - made reaction extracellular (really periplasmic); JAB 04/05/14 - added when investigating accumulation of gluconate/2-ketogluconate and problems in aerobic metabolism, 2-dehydro-d-gluconate to gluconate direction preferred according to Metacyc, listed as multidirectional in KEGG'\n",
    "\n",
    "model.reactions.rPY00214.id = 'rxn06971' #line 78/79\n",
    "model.reactions.rxn06971.name = 'CDP-diacylglycerol:choline O-phosphatidyltransferase'\n",
    "model.reactions.rxn06971.add_metabolites({model.metabolites.cpd00067_c: 1.0,})\n",
    "model.reactions.rxn06971.annotation['ec-code'] ='2.7.8.24'\n",
    "model.reactions.rxn06971.notes['NOTES'] = '*LJD  12/14/18 - replaced rPY00214 with equivalent reaction in model seed. PY 04/14/14 - PAO1 can produce phosphatidylcholine (aka PC, lecithin) when grown on rich media and minimal media when choline is provided, but cannot produce PC on minimal media when choline is not provided. PA3857 is the phosphatidylchoine synthase (Pcs) gene. There also appears to be a PmtA gene (PA0798), which in other bacteria, produces PC from the triple methylation of phosphatidylethanolamine, but this gene seems to be nonfunctional in PAO1. Reaction is essentially rxn06971 from SEED master list.'\n",
    "\n",
    "model.reactions.rAB00001.id = 'rxn00979' #line 11/12\n",
    "model.reactions.rxn00979.notes['NOTES'] = '*LJD 12/12/18 - replaced ASB reaction rAB00001 with identical model seed reaction. ASB 05/30/14 - Added to fix glycolaldehyde usage, gene is based off of BLAST results and may not be correct, see paper Dagley, 1961 Biochem J'\n",
    "\n",
    "model.reactions.rJB00258.id = 'rxn01805' #line 29/30 - weird adding metabolites because wanted to change order of rxn\n",
    "model.reactions.rxn01805.add_metabolites({\n",
    "    model.metabolites.cpd00743_c: -2.0,\n",
    "    model.metabolites.cpd00416_c:2.0,\n",
    "    model.metabolites.cpd00001_c:2.0,})\n",
    "model.reactions.rxn01805.notes['NOTES'] = '*LJD 12/12/19 - replaced rJB00258 with rxn01805 from seed. Flipped order and added E.C number and kegg reaction ID. JAB 02/08/14 - added from MetaCyc/lit for Biolog validation (Itaconate), rxn favored in forward dir but leaving rev for now'\n",
    "model.reactions.rxn01805.references = {'REFERENCES':'Cooper, R. a, & Kornberg, H. L. (1964). The utilization of itaconate by Pseudomonas sp. The Biochemical Journal, 91(1), 82–91. Retrieved from http://www.pubmedcentral.nih.gov/articlerender.fcgi?artid=1202819&tool=pmcentrez&rendertype=abstract'}\n",
    "model.reactions.rxn01805.annotation['seed.reaction'] = 'rxn01805'\n",
    "model.reactions.rxn01805.annotation['ec-code'] = '4.2.1.56'\n",
    "model.reactions.rxn01805.annotation['bigg.metabolite'] = 'R02491'\n",
    "\n",
    "#Remove reactions\n",
    "model.reactions.EX_cJB00034_e.remove_from_model() #line4\n",
    "model.reactions.EX_cPY00016_e.remove_from_model() #line6\n",
    "model.reactions.EX_cPY00116_e.remove_from_model() #line8\n",
    "\n",
    "#Add reactions\n",
    "reaction1 = Reaction('EX_cpd00212_e') #line 5\n",
    "reaction1.name = 'EX L-Sorbose'\n",
    "reaction1.subsystem = 'Exchange'\n",
    "reaction1.lower_bound = 0\n",
    "reaction1.upper_bound = 0\n",
    "reaction1.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction1.add_metabolites({model.metabolites.cpd00212_e: -1.0,})\n",
    "reaction1.notes['NOTES'] = '*LJD 12/12/18 - Replaced EX_cJB00034 with seed id.'\n",
    "model.add_reactions([reaction1])\n",
    "\n",
    "reaction2 = Reaction('EX_cpd00138_e') #line 7\n",
    "reaction2.name = 'EX D-Mannose'\n",
    "reaction2.subsystem = 'Exchange'\n",
    "reaction2.lower_bound = 0\n",
    "reaction2.upper_bound = 0\n",
    "reaction2.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction2.add_metabolites({model.metabolites.cpd00138_e: -1.0,})\n",
    "reaction2.notes['NOTES'] = '*LJD 12/12/18 - Replaced redundant PY exchange with exchange with seed ID'\n",
    "model.add_reactions([reaction2])\n",
    "\n",
    "reaction3 = Reaction('EX_cpd03587_e') #line 9\n",
    "reaction3.name = 'EX kdo2-lipid a'\n",
    "reaction3.subsystem = 'Exchange'\n",
    "reaction3.lower_bound = 0\n",
    "reaction3.upper_bound = 0\n",
    "reaction3.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction3.add_metabolites({model.metabolites.cpd03587_e: -1.0,})\n",
    "reaction3.notes['NOTES'] = '*LJD 12/12/18 - Replaced cPY00016 with model seed id'\n",
    "model.add_reactions([reaction3])\n",
    "\n",
    "#_____________________________________________________________________________________________________\n",
    "#Make changes based on Literature-based updates\n",
    "\n",
    "#Metabolites\n",
    "\n",
    "#line 2 of LJD\n",
    "cpd00477_e = Metabolite('cpd00477_e',\n",
    "                        formula='C7H9NO5',\n",
    "                        name='N-Acetyl-L-glutamate',\n",
    "                        compartment='extracellular',\n",
    "                        charge = -2)\n",
    "model.add_metabolites([cpd00477_e])\n",
    "model.metabolites.cpd00477_e.notes= {'NOTES':'LJD 1/31/20 - added extracellular version of metabolite'}\n",
    "\n",
    "#Line 5 of DDP\n",
    "cpd02175_c = Metabolite('cpd02175_c',\n",
    "                        formula='C5H9NO3',\n",
    "                        name='cis-4-Hydroxy-D-proline',\n",
    "                        compartment='cytosol',\n",
    "                        charge=0)\n",
    "model.add_metabolites([cpd02175_c])\n",
    "model.metabolites.cpd02175_c.annotation={'kegg.compound':'C03440','MetaCyc':'CIS-4-HYDROXY-D-PROLINE','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02175_c.notes= {'NOTES':'DDP 3/3/20 - Added because present in rxn02360 and rLD02946, which were added by LJD'}\n",
    "\n",
    "#Line 6 of DDP\n",
    "met12 = Metabolite('cpd02626_c',\n",
    "                   formula = 'C5H6NO3',\n",
    "                   name = '1-Pyrroline-4-hydroxy-2-carboxylate',\n",
    "                   compartment = 'cytosol',\n",
    "                   charge = -1)\n",
    "model.add_metabolites([met12])\n",
    "model.metabolites.cpd02626_c.annotation = {'kegg.compound':'C04282','MetaCyc':'CPD-1128','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02626_c.notes = {'NOTES':'DDP 3/3/20 - Added because present in rxn01635 and rLD02946, which were added by LJD'}\n",
    "\n",
    "#Line 7 of DDP\n",
    "met13 = Metabolite('cpd00340_c',\n",
    "                   formula = 'C5H5O4',\n",
    "                   name = '2,5-Dioxopentanoate',\n",
    "                   compartment = 'cytosol',\n",
    "                   charge = -1)\n",
    "model.add_metabolites([met13])\n",
    "model.metabolites.cpd00340_c.annotation = {'bigg.metabolite':'25dop','kegg.compound':'C00433','MetaCyc':'CPD-654','iAbaylyiv4':'CPD-654','iRS1563':'M_C00433_c','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd00340_c.notes = {'NOTES':'DDP 3/3/20 - Added because present in rxn01635 and rxn00196 which were added by LJD'}\n",
    "\n",
    "#Line 8 of DDP\n",
    "met14 = Metabolite('cpd01715_c',formula = 'C7H13NO3',name = '3-Dehydrocarnitine',compartment = 'cytosol',charge = 0)\n",
    "model.add_metabolites([met14])\n",
    "model.metabolites.cpd01715_c.annotation = {'kegg.compound':'C02636','sbo': 'SBO:0000247','MetaCyc':'3-DE-H-CARNITINE'}\n",
    "model.metabolites.cpd01715_c.notes = {'NOTES':'DDP 3/3/20 - Added because present in rxn01725 and rLD01726 which were added by LJD'}\n",
    "\n",
    "#line 21 of LJD\n",
    "cpd02507_c = Metabolite('cpd02507_c',\n",
    "                        formula='C8H5O7',\n",
    "                        name='5-Carboxy-2-oxohept-3-enedioate',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -3)\n",
    "model.add_metabolites([cpd02507_c])\n",
    "model.metabolites.cpd02507_c.notes = {'NOTES':'LJD 2/5/20 - added missing metabolites for p-hydroxy phenyl acetic acid metabolism'}\n",
    "model.metabolites.cpd02507_c.annotation = {'kegg.compound':'C04052','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02507_c.annotation['bigg.metabolite'] = '5cohe'\n",
    "\n",
    "#line 22 of LJD\n",
    "cpd03326_c= Metabolite('cpd03326_c',\n",
    "                        formula='C7H6O5',\n",
    "                        name='2-Hydroxyhepta-2,4-dienedioate',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -2)\n",
    "model.add_metabolites([cpd03326_c])\n",
    "model.metabolites.cpd03326_c.notes = {'NOTES':'LJD 2/5/20 - added missing metabolites for p-hydroxy phenyl acetic acid metabolism'}\n",
    "model.metabolites.cpd03326_c.annotation = {'kegg.compound':'C05600','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd03326_c.annotation['bigg.metabolite']='2hh24dd'\n",
    "\n",
    "#line 23 of LJD\n",
    "cpd03708_c= Metabolite('cpd03708_c',\n",
    "                        formula='C7H8O6',\n",
    "                        name='2,4-Dihydroxyhept-2-enedioate',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -2)\n",
    "model.add_metabolites([cpd03708_c])\n",
    "model.metabolites.cpd03708_c.notes = {'NOTES':'LJD 2/5/20 - added missing metabolites for p-hydroxy phenyl acetic acid metabolism'}\n",
    "model.metabolites.cpd03708_c.annotation = {'bigg.metabolite':'24dhhed','sbo': 'SBO:0000247','kegg.compound':'C06201','MetaCyc':'CPD-15125','TS_Athaliana':'met311_c','iMA945':'24dhh2ed'}\n",
    "\n",
    "#line 33 of LJD\n",
    "cpd14700_c= Metabolite('cpd14700_c',\n",
    "                        formula='C16H25N4O8P2S',\n",
    "                        name='2-Methyl-1-hydroxypropyl-TPP',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -1)\n",
    "model.add_metabolites([cpd14700_c])\n",
    "model.metabolites.cpd14700_c.annotation = {'kegg.compound':'C15976','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd14700_c.annotation['bigg.metabolite']='2mhop'\n",
    "\n",
    "#line 34 of LJD\n",
    "cpd02700_c= Metabolite('cpd02700_c',\n",
    "                        formula='C12H23NO2S2',\n",
    "                        name=' S-(2-Methylpropionyl)-dihydrolipoamide',\n",
    "                        compartment='cytosol',\n",
    "                        charge = 0)\n",
    "model.add_metabolites([cpd02700_c])\n",
    "model.metabolites.cpd02700_c.annotation = {'bigg.metabolite':'2mpdhl','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02700_c.annotation['kegg.compound']='C04424'\n",
    "model.metabolites.cpd02700_c.annotation['MetaCyc']='CPD-281'\n",
    "\n",
    "#line 35 of LJD\n",
    "cpd02187_c= Metabolite('cpd02187_c',\n",
    "                        formula='C25H36N7O17P3S',\n",
    "                        name='Methacrylyl-CoA',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -4)\n",
    "model.add_metabolites([cpd02187_c])\n",
    "model.metabolites.cpd02187_c.annotation = {'bigg.metabolite':'2mp2coa','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02187_c.annotation['kegg.compound']='C03460'\n",
    "model.metabolites.cpd02187_c.annotation['MetaCyc']='METHACRYLYL-COA'\n",
    "\n",
    "#line 36 of LJD\n",
    "cpd03572_c= Metabolite('cpd03572_c',\n",
    "                        formula='C25H38N7O18P3S',\n",
    "                        name='(S)-3-Hydroxyisobutyryl-CoA',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -4)\n",
    "model.add_metabolites([cpd03572_c])\n",
    "model.metabolites.cpd03572_c.annotation = {'bigg.metabolite':'3hibutcoa','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd03572_c.annotation['kegg.compound']='C06000'\n",
    "model.metabolites.cpd03572_c.annotation['MetaCyc']='CPD-12174'\n",
    "\n",
    "#line 37 of LJD\n",
    "cpd02096_c= Metabolite('cpd02096_c',\n",
    "                        formula='C4H9NO2',\n",
    "                        name='L-3-Amino-isobutyrate',\n",
    "                        compartment='cytosol',\n",
    "                        charge = 0)\n",
    "model.add_metabolites([cpd02096_c])\n",
    "model.metabolites.cpd02096_c.annotation = {'bigg.metabolite':'3aib','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd02096_c.annotation['kegg.compound']='C03284'\n",
    "model.metabolites.cpd02096_c.annotation['MetaCyc']='CPD-466'\n",
    "\n",
    "#Reactions\n",
    "\n",
    "#line 2 of LJD\n",
    "reaction101 = Reaction('EX_cpd00477_e')\n",
    "reaction101.name = 'EX N-Acetyl-L-glutamate'\n",
    "reaction101.subsystem = 'Exchange'\n",
    "reaction101.lower_bound = 0\n",
    "reaction101.upper_bound = 0\n",
    "reaction101.annotation = {'sbo': 'SBO:0000627'}\n",
    "reaction101.add_metabolites({cpd00477_e: -1.0,})\n",
    "model.add_reactions([reaction101])\n",
    "\n",
    "model.metabolites.cpd00477_e.annotation = {'sbo': 'SBO:0000247','bigg.metabolite': 'acglu','biocyc': 'META:ACETYL-GLU','chebi': ['CHEBI:12575','CHEBI:17533','CHEBI:21549','CHEBI:21552','CHEBI:44335','CHEBI:44337','CHEBI:64040','CHEBI:7150'],'hmdb': 'HMDB01138','kegg.compound': 'C00624','metanetx.chemical': 'MNXM730','seed.compound':'cpd00477','unipathway.compound': 'UPC00624'}\n",
    "model.metabolites.cpd00477_e.charge = -2\n",
    "\n",
    "#line 3 of LJD\n",
    "reaction102 = Reaction('rLD05146')\n",
    "reaction102.name = 'N-Acetyl-L-glutamate-ABC transport'\n",
    "reaction102.subsystem = 'ABC transporter N-Acetyl-L-glutamate'\n",
    "reaction102.lower_bound = 0\n",
    "reaction102.upper_bound = 1000\n",
    "reaction102.annotation = {'ec-code':'7.4.2.1','sbo': 'SBO:0000185', 'KEGG Subsystem':'Transport'}\n",
    "reaction102.gene_reaction_rule = 'PA14_46950 and PA14_46930 and PA14_46920 and PA14_46910'\n",
    "reaction102.notes = {'NOTES':'LJD 1/30/20 - From referenced paper, identified ABC transport that should impact entry of N-Acetyl-Glutamic Acid into the cell in PAO1. Found an orthologous PA14 gene for this transporters. Carried over the rest of the genes in the complex from rxn05155. Did not include second complex from rxn05155 because I could not find evidence to support. Ec-code from similar modelSEED reaction rxn05155'}\n",
    "reaction102.references = {'REFERENCES':'Johnson DA, Tetu SG, Phillippy K, Chen J, Ren Q, Paulsen IT (2008) High-Throughput Phenotypic Characterization of Pseudomonas aeruginosa Membrane Transport Genes. Plos Genetics. 4(10): e1000211.'}\n",
    "reaction102.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd00002_c: -1.0,\n",
    "    model.metabolites.cpd00477_e: -1.0,\n",
    "    model.metabolites.cpd00067_c: 1.0,\n",
    "    model.metabolites.cpd00477_c: 1.0,\n",
    "    model.metabolites.cpd00008_c: 1.0,\n",
    "    model.metabolites.cpd00009_c: 1.0,})\n",
    "model.add_reactions([reaction102])\n",
    "\n",
    "#Line 4 of LJD\n",
    "reaction103 = Reaction('rLD05298')\n",
    "reaction103.name = 'Na+/N-Acetyl-L-glutamate symport'\n",
    "reaction103.subsystem = 'Transport Extracellular'\n",
    "reaction103.lower_bound = -1000\n",
    "reaction103.upper_bound = 1000\n",
    "reaction103.annotation = {'ec-code':'7.4.2.1','sbo': 'SBO:0000185','KEGG Subsystem':'Transport'}\n",
    "reaction103.gene_reaction_rule = 'PA14_23160'\n",
    "reaction103.notes = {'NOTES':'LJD 1/30/20 - From referenced paper, identified sodium transport systems that should both impact entry of N-Acetyl-Glutamic Acid into the cell in PAO1. Found an orthologous PA14 gene for this transporters.'}\n",
    "reaction103.references = {'REFERENCES':'Johnson DA, Tetu SG, Phillippy K, Chen J, Ren Q, Paulsen IT (2008) High-Throughput Phenotypic Characterization of Pseudomonas aeruginosa Membrane Transport Genes. Plos Genetics. 4(10): e1000211.'}\n",
    "reaction103.add_metabolites({\n",
    "    model.metabolites.cpd00971_e: -1.0,\n",
    "    model.metabolites.cpd00477_e: -1.0,\n",
    "    model.metabolites.cpd00477_c: 1.0,\n",
    "    model.metabolites.cpd00971_c: 1.0,})\n",
    "model.add_reactions([reaction103])\n",
    "\n",
    "#Line 5\n",
    "reaction104 = Reaction('rxn12634')\n",
    "reaction104.name = 'methionyl aminopeptidase'\n",
    "reaction104.lower_bound = 0\n",
    "reaction104.upper_bound = 1000\n",
    "reaction104.annotation = {'ec-code':'3.4.11.18','sbo': 'SBO:0000176','seed.reaction':'rxn12634'}\n",
    "reaction104.gene_reaction_rule = 'PA14_17050 or PA14_28590'\n",
    "reaction104.notes = {'NOTES':'LJD 1/21/20 - Adding this reaction to the model results in growth on Glycyl-L-Glutamic Acid; PA14_28590 is putative. Assuming OR GPR. KEGG and PDB both support GPR'}\n",
    "reaction104.references = {'REFERENCES':'Ouidir T, Jarnier F, Cosette P, Jouenne T, Hardouin J (2015) Characterization of N-terminal protein modifications in Pseudomonas aeruginosa PA14. Journal of Proteomics 114:214-225'}\n",
    "reaction104.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd11592_c: -1.0,\n",
    "    model.metabolites.cpd00023_c: 1.0,\n",
    "    model.metabolites.cpd00033_c: 1.0,})\n",
    "model.add_reactions([reaction104])\n",
    "\n",
    "#Line 6\n",
    "reaction105 = Reaction('rxn12638')\n",
    "reaction105.name = 'methionyl aminopeptidase'\n",
    "reaction105.lower_bound = 0\n",
    "reaction105.upper_bound = 1000\n",
    "reaction105.annotation = {'ec-code':'3.4.11.18','sbo': 'SBO:0000176','seed.reaction':'rxn12638'}\n",
    "reaction105.gene_reaction_rule = 'PA14_17050 or PA14_28590'\n",
    "reaction105.notes = {'NOTES':'LJD 1/21/20 - Adding this reaction to the model results in growth on Glycyl-L-Proline; PA14_28590 is putative. Assuming OR GPR. KEGG and PDB both support GPR'}\n",
    "reaction105.references = {'REFERENCES':'Ouidir T, Jarnier F, Cosette P, Jouenne T, Hardouin J (2015) Characterization of N-terminal protein modifications in Pseudomonas aeruginosa PA14. Journal of Proteomics 114:214-225'}\n",
    "reaction105.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd11588_c: -1.0,\n",
    "    model.metabolites.cpd00129_c: 1.0,\n",
    "    model.metabolites.cpd00033_c: 1.0,})\n",
    "model.add_reactions([reaction105])\n",
    "\n",
    "#Line 7\n",
    "reaction106 = Reaction('rxn02360')\n",
    "reaction106.name = 'trans-4-Hydroxy-L-proline 2-epimerase'\n",
    "reaction106.subsystem = 'Amino Acid Degradation'\n",
    "reaction106.lower_bound = -1000\n",
    "reaction106.upper_bound = 1000\n",
    "reaction106.annotation = {'ec-code':'5.1.1.8','seed.reaction':'rxn02360','sbo': 'SBO:0000176','KEGG Subsystem':'Arginine and proline metabolism'}\n",
    "reaction106.gene_reaction_rule = 'PA14_47840'\n",
    "reaction106.notes = {'NOTES':'LJD 1/31/20 - Found literature evidence of catabolism of trans-4-hydroxy-l-proline and addressing biolog need. Added 4 reactions to get from this carbon source to 2-oxoglutarate. Genes are PA14 orthologs from reference'}\n",
    "reaction106.references = {'REFERENCES':'https://www.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.000300#tab2'}\n",
    "reaction106.add_metabolites({\n",
    "    model.metabolites.cpd00851_c: -1.0,\n",
    "    model.metabolites.cpd02175_c: 1.0,})\n",
    "model.add_reactions([reaction106])\n",
    "\n",
    "#Line 8\n",
    "reaction107 = Reaction('rLD02946')\n",
    "reaction107.name = 'cis-4-Hydroxy-D-proline:FAD oxidoreductase'\n",
    "reaction107.subsystem = 'Amino Acid Degradation'\n",
    "reaction107.lower_bound = 0\n",
    "reaction107.upper_bound = 1000\n",
    "reaction107.annotation = {'ec-code':'1.5.99.-','sbo': 'SBO:0000176', 'KEGG Subsystem':'Arginine and proline metabolism'}\n",
    "reaction107.gene_reaction_rule = '(PA14_47850 and PA14_47860)'\n",
    "reaction107.notes = {'NOTES':'LJD 1/31/20 - Found literature evidence of catabolism of trans-4-hydroxy-l-proline and addressing biolog need. Reaction converting hydroxy-D-proline to 1-pyrroline-4-hydroxy-2-carboxylate has genetic evidence and an enzyme family, but the exact reaction is unknown. Based off of EC 1.5.99.13, created this reaction and used FAD as a proton acceptor. Made irreversible similar to rxn02946 in SEED which has an enzyme without PA14 support and produces hydrogen peroxide'}\n",
    "reaction107.references = {'REFERENCES':'https://www.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.000300#tab2'}\n",
    "reaction107.add_metabolites({\n",
    "    model.metabolites.cpd02175_c: -1.0,\n",
    "    model.metabolites.cpd00015_c: -1.0,\n",
    "    model.metabolites.cpd02626_c: 1.0,\n",
    "    model.metabolites.cpd00067_c: 1.0,\n",
    "    model.metabolites.cpd00982_c: 1.0,})\n",
    "model.add_reactions([reaction107])\n",
    "\n",
    "#Line  9\n",
    "reaction108 = Reaction('rxn01635')\n",
    "reaction108.name = '1-Pyrroline-4-hydroxy-2-carboxylate aminohydrolase (decyclizing)'\n",
    "reaction108.subsystem = 'Amino Acid Degradation'\n",
    "reaction108.lower_bound = 0\n",
    "reaction108.upper_bound = 1000\n",
    "reaction108.annotation = {'seed.reaction':'rxn01635','ec-code':'3.5.4.22','sbo': 'SBO:0000176','KEGG Subsystem':'Arginine and proline metabolism'}\n",
    "reaction108.gene_reaction_rule = 'PA14_48000'\n",
    "reaction108.notes = {'NOTES':'LJD 1/31/20 - Found literature evidence of catabolism of trans-4-hydroxy-l-proline and addressing biolog need. Added 4 reactions to get from this carbon source to 2-oxoglutarate. Genes are PA14 orthologs from reference'}\n",
    "reaction108.references = {'REFERENCES':'https://www.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.000300#tab2'}\n",
    "reaction108.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd02626_c: -1.0,\n",
    "    model.metabolites.cpd00067_c: -1.0,\n",
    "    model.metabolites.cpd00013_c: 1.0,\n",
    "    model.metabolites.cpd00340_c: 1.0,})\n",
    "model.add_reactions([reaction108])\n",
    "\n",
    "\n",
    "#Line 10\n",
    "reaction109 = Reaction('rxn00196')\n",
    "reaction109.name = '2,5-dioxopentanoate:NADP+ 5-oxidoreductase'\n",
    "reaction109.subsystem = 'Amino Acid Degradation'\n",
    "reaction109.lower_bound = -1000\n",
    "reaction109.upper_bound = 1000\n",
    "reaction109.annotation = {'seed.reaction':'rxn00196','ec-code':'1.2.1.26','sbo': 'SBO:0000176', 'KEGG Subsystem':'Pentose and glucuronate interconversions'}\n",
    "reaction109.gene_reaction_rule = 'PA14_48010 or PA14_36050'\n",
    "reaction109.notes = {'NOTES':'LJD 1/31/20 - Found literature evidence of catabolism of trans-4-hydroxy-l-proline and addressing biolog need. Added 4 reactions to get from this carbon source to 2-oxoglutarate. Genes are PA14 orthologs from reference and other genes associated with Ec-code in Kegg'}\n",
    "reaction109.references = {'REFERENCES':'https://www.microbiologyresearch.org/content/journal/micro/10.1099/mic.0.000300#tab2'}\n",
    "reaction109.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd00006_c: -1.0,\n",
    "    model.metabolites.cpd00340_c: -1.0,\n",
    "    model.metabolites.cpd00005_c: 1.0,\n",
    "    model.metabolites.cpd00024_c: 1.0,\n",
    "    model.metabolites.cpd00067_c: 2.0,})\n",
    "model.add_reactions([reaction109])\n",
    "\n",
    "#Line 11\n",
    "reaction110 = Reaction('rxn00178')\n",
    "reaction110.name = 'Acetyl-CoA:acetyl-CoA C-acetyltransferase'\n",
    "reaction110.lower_bound = -1000\n",
    "reaction110.upper_bound = 1000\n",
    "reaction110.annotation = {'seed.reaction':'rxn00178','ec-code':'2.3.1.9','sbo': 'SBO:0000176'}\n",
    "reaction110.gene_reaction_rule = 'PA14_13090 or PA14_17880 or PA14_31530 or PA14_38630 or PA14_42090 or PA14_63250'\n",
    "reaction110.references = {'REFERENCES':'Arevalo-Ferro Catalina, Hentzer M, Reil G, Gorg A, Kjelleberg S, Givskov M, Riedel K, Eberl L (2003) Identification of quorum-sensing regulated proteins in the opportunistic pathogen Pseudomonas aeruginosa by proteomics. Environmental Microbiology. 5(12):1350-1369'}\n",
    "reaction110.add_metabolites({\n",
    "    model.metabolites.cpd00022_c: -2.0,\n",
    "    model.metabolites.cpd00010_c: 1.0,\n",
    "    model.metabolites.cpd00279_c: 1.0,})\n",
    "model.add_reactions([reaction110])\n",
    "\n",
    "#Line 13\n",
    "reaction112 = Reaction('rxn01725')\n",
    "reaction112.name = 'Carnitine:NAD+ 3-oxidoreductase'\n",
    "reaction112.subsystem = 'Carnitine Metabolism'\n",
    "reaction112.lower_bound = -1000\n",
    "reaction112.upper_bound = 1000\n",
    "reaction112.annotation = {'seed.reaction':'rxn01725','ec-code':'1.1.1.108','sbo': 'SBO:0000176'}\n",
    "reaction112.gene_reaction_rule = 'PA14_71140'\n",
    "reaction112.notes = {'NOTES':'LJD 2/3/20 - Added based on literature evidence'}\n",
    "reaction112.references = {'REFERENCES':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/'}\n",
    "reaction112.add_metabolites({\n",
    "    model.metabolites.cpd00003_c: -1.0,\n",
    "    model.metabolites.cpd00266_c: -1.0,\n",
    "    model.metabolites.cpd00004_c: 1.0,\n",
    "    model.metabolites.cpd00067_c: 1.0,\n",
    "    model.metabolites.cpd01715_c: 1.0,})\n",
    "model.add_reactions([reaction112])\n",
    "\n",
    "#Line 14\n",
    "met10 = Metabolite('cLD00540_c',\n",
    "                 formula = 'C26H42N8O17P3S',\n",
    "                 name = 'BET-CoA',\n",
    "                 compartment = 'cytosol',\n",
    "                 charge = -3)\n",
    "model.add_metabolites([met10])\n",
    "model.metabolites.cLD00540_c.annotation = {'bigg.metabolite':'glybcoa[c]','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cLD00540_c.references = {'References':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/'}\n",
    "model.metabolites.cLD00540_c.notes = {'NOTES':'LJD 2/3/20 - Found literature evidence and evidence from putida model that Pseudomonas can convert 3-dehydrocarnitine to BET via dehydrocarnitine-coA and betaine-coA'}\n",
    "\n",
    "#Line 15\n",
    "reaction114 = Reaction('rLD01726')\n",
    "reaction114.name = 'BKACE CoA transferase'\n",
    "reaction114.subsystem = 'Carnitine Metabolism'\n",
    "reaction114.lower_bound = -1000\n",
    "reaction114.upper_bound = 1000\n",
    "reaction114.annotation = {'sbo': 'SBO:0000176'}\n",
    "reaction114.gene_reaction_rule = ''\n",
    "reaction114.notes = {'NOTES':'LJD 2/3/20 - Found literature evidence and evidence from putida model that Pseudomonas can convert 3-dehydrocarnitine to BET. Putida model has an extra intermediate reaction going from 3-dhc -> 3-dhc-coA -> bet-coA -> bet. Bastard et al 2014 (Nat Chem Biology) show PAO1 can go from 3-dhc -> bet-coA -> bet, which is what I included. These reactions do lack enzymatic and genetic evidence from that paper. Consider shifting to putida approach'}\n",
    "reaction114.references = {'REFERENCES':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/, https://www.ncbi.nlm.nih.gov/pubmed/19406895, https://www.nature.com/articles/nchembio.1387/figures/5'}\n",
    "reaction114.add_metabolites({\n",
    "    model.metabolites.cpd01715_c: -1.0,\n",
    "    model.metabolites.cpd00022_c: -1.0,\n",
    "    model.metabolites.cpd00142_c: 1.0,\n",
    "    model.metabolites.cLD00540_c: 1.0,\n",
    "    model.metabolites.cpd00067_c: 1.0})\n",
    "model.add_reactions([reaction114])\n",
    "\n",
    "#Line 16\n",
    "reaction115 = Reaction('rLD01727')\n",
    "reaction115.name = 'BET thiolase'\n",
    "reaction115.subsystem = 'Carnitine Metabolism'\n",
    "reaction115.lower_bound = -1000\n",
    "reaction115.upper_bound = 1000\n",
    "reaction115.annotation = {'sbo':'SBO:0000176'}\n",
    "reaction115.notes = {'NOTES':'LJD 2/3/20 - Found literature evidence and evidence from putida model that Pseudomonas can convert 3-dehydrocarnitine to BET. Putida model has an extra intermediate reaction going from 3-dhc -> 3-dhc-coA -> bet-coA -> bet. Bastard et al 2014 (Nat Chem Biology) show PAO1 can go from 3-dhc -> bet-coA -> bet, which is what I included. These reactions do lack enzymatic and genetic evidence from that paper. Consider shifting to putida approach'}\n",
    "reaction115.references = {'REFERENCES':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/, https://www.ncbi.nlm.nih.gov/pubmed/19406895, https://www.nature.com/articles/nchembio.1387/figures/5'}\n",
    "reaction115.add_metabolites({\n",
    "    model.metabolites.cLD00540_c: -1.0,\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd00540_c: 1.0,\n",
    "    model.metabolites.cpd00010_c: 1.0,})\n",
    "model.add_reactions([reaction115])\n",
    "\n",
    "#Line 17\n",
    "model.reactions.rxn02028.gene_reaction_rule = '(PA14_71420 and PA14_71410 and PA14_24290)'\n",
    "model.reactions.rxn02028.subsystem = 'Carnitine Metabolism'\n",
    "model.reactions.rxn02028.annotation['KEGG Subsystem'] = 'Glycine, serine and threonine metabolism'\n",
    "model.reactions.rxn02028.notes = {'NOTES':'LJD 2/3/20 - Updated reaction GPR based on literature evidence. Two genes both found to be required for growth on BET but not on DMG. Kept PA14_24290 from past update'}\n",
    "model.reactions.rxn02028.references = {'REFERENCES':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/ https://jb.asm.org/content/190/8/2690.full'}\n",
    "\n",
    "#Line 18\n",
    "reaction118 = Reaction('rxn12739')\n",
    "reaction118.name = 'dimethylglycine dehydrogenase'\n",
    "reaction118.subsystem = 'Carnitine Metabolism'\n",
    "reaction118.lower_bound = -1000\n",
    "reaction118.upper_bound = 1000\n",
    "reaction118.annotation = {'seed.reaction':'rxn12739','ec-code':'1.5.8.4','sbo': 'SBO:0000176'}\n",
    "reaction118.gene_reaction_rule = 'PA14_71280 and PA14_71260'\n",
    "reaction118.notes = {'NOTES':'LJD 2/3/20 - Genetic evidence supports conversion of DMG to sarcosine; No enzymatic evidence yet; Unclear what substrates are, so I favored FAD/FADH2 from putida model over O2/H2O2 in existing SEED reactions'}\n",
    "reaction118.references = {'REFERENCES':'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5806209/ https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2293255/'}\n",
    "reaction118.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd00015_c: -1.0,\n",
    "    model.metabolites.cpd00756_c: -1.0,\n",
    "    model.metabolites.cpd00055_c: 1.0,\n",
    "    model.metabolites.cpd00183_c: 1.0,\n",
    "    model.metabolites.cpd00982_c: 1.0,})\n",
    "model.add_reactions([reaction118])\n",
    "\n",
    "#Line 24 of LJD\n",
    "reaction124 = Reaction('rxn03040')\n",
    "reaction124.name = '5-Carboxymethyl-2-hydroxymuconate delta2,delta4-2-oxo,delta3-isomerase'\n",
    "reaction124.subsystem = 'Tyrosine Metabolism'\n",
    "reaction124.lower_bound = -1000\n",
    "reaction124.upper_bound = 1000\n",
    "reaction124.annotation = {'seed.reaction':'rxn03040','ec-code':'5.3.3.10','KEGG Subsystem':'Tyrosine Metabolism','sbo': 'SBO:0000176'}\n",
    "reaction124.gene_reaction_rule = 'PA14_10610 or PA14_39100'\n",
    "reaction124.notes = {'NOTES':'LJD 2/5/20 - added kegg pathway for p-hydroxy phenyl acetic acid metabolism and associated GPRs; metacyc says irreversible but SEED said reversible. Sticking with SEED for now; GPR OR relationship'}\n",
    "reaction124.add_metabolites({\n",
    "    model.metabolites.cpd02572_c: -1.0,\n",
    "    model.metabolites.cpd02507_c: 1.0,})\n",
    "model.add_reactions([reaction124])\n",
    "\n",
    "#Line 25 of LJD\n",
    "reaction125 = Reaction('rxn03041')\n",
    "reaction125.name = '5-oxopent-3-ene-1,2,5-tricarboxylate carboxy-lyase'\n",
    "reaction125.subsystem = 'Tyrosine Metabolism'\n",
    "reaction125.lower_bound = 0\n",
    "reaction125.upper_bound = 1000\n",
    "reaction125.annotation = {'seed.reaction':'rxn03041','ec-code':'4.1.1.68','sbo': 'SBO:0000176', 'KEGG Subsystem':'Tyrosine Metabolism'}\n",
    "reaction125.gene_reaction_rule = 'PA14_10640 or PA14_10650'\n",
    "reaction125.notes = {'NOTES':'LJD 2/5/20 - added kegg pathway for p-hydroxy phenyl acetic acid metabolism and associated GPRs; seed says irreversible; Both genes are hypothetically related to EC 4.1.1.68, but left GPR as an OR relationship because I could not find evidence of them working together in KEGG or the pseudomonas database'}\n",
    "reaction125.add_metabolites({\n",
    "    model.metabolites.cpd00067_c: -1.0,\n",
    "    model.metabolites.cpd02507_c: -1.0,\n",
    "    model.metabolites.cpd00011_c: 1.0,\n",
    "    model.metabolites.cpd03326_c: 1.0,})\n",
    "model.add_reactions([reaction125])\n",
    "\n",
    "\n",
    "#Line 26\n",
    "reaction126 = Reaction('rxn04706')\n",
    "reaction126.name = '2,4-dihydroxyhept-2-enedioate hydro-lyase'\n",
    "reaction126.subsystem = 'Tyrosine Metabolism'\n",
    "reaction126.lower_bound = -1000\n",
    "reaction126.upper_bound = 1000\n",
    "reaction126.annotation = {'seed.reaction':'rxn04706','ec-code':'4.2.1.-','sbo': 'SBO:0000176','KEGG Subsystem':'Tyrosine Metabolism'}\n",
    "reaction126.gene_reaction_rule = 'PA14_10590'\n",
    "reaction126.notes = {'NOTES':'LJD 2/5/20 - added kegg pathway for p-hydroxy phenyl acetic acid metabolism and associated GPRs; seed says reversible'}\n",
    "reaction126.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd03326_c: -1.0,\n",
    "    model.metabolites.cpd03708_c: 1.0,})\n",
    "model.add_reactions([reaction126])\n",
    "\n",
    "#Line 27 of LJD\n",
    "reaction127 = Reaction('rxn01203')\n",
    "reaction127.name = '2,4-dihydroxyhept-2-ene-1,7-dioic acid aldolase'\n",
    "reaction127.subsystem = 'Tyrosine Metabolism'\n",
    "reaction127.lower_bound = -1000\n",
    "reaction127.upper_bound = 0\n",
    "reaction127.annotation = {'seed.reaction':'rxn01203','ec-code':'4.1.2.52','sbo': 'SBO:0000176','KEGG Subsystem':'Tyrosine Metabolism'}\n",
    "reaction127.gene_reaction_rule = 'PA14_10570'\n",
    "reaction127.notes = {'NOTES':'LJD 2/5/20 - added kegg pathway for p-hydroxy phenyl acetic acid metabolism and associated GPRs; metacyc says reversible while seed has it written as irreversible and flipped; kept with seed and note bounds reflect the flipping'}\n",
    "reaction127.add_metabolites({\n",
    "    model.metabolites.cpd00199_c: -1.0,\n",
    "    model.metabolites.cpd00020_c: -1.0,\n",
    "    model.metabolites.cpd03708_c: 1.0,})\n",
    "model.add_reactions([reaction127])\n",
    "\n",
    "#Line 28 of LJD\n",
    "model.reactions.rxn00670.lower_bound = -1000\n",
    "model.reactions.rxn00670.upper_bound = 1000\n",
    "model.reactions.rxn00670.notes['NOTES'] = 'LJD 2/3/20 - opened reaction bounds'\n",
    "\n",
    "#Line 29 of LJD\n",
    "model.reactions.rxn00985.lower_bound = -1000\n",
    "model.reactions.rxn00985.upper_bound = 1000\n",
    "model.reactions.rxn00985.notes['NOTES'] = 'LJD 2/3/20 - opened reaction bounds. No KEGG PA14 evidence for 2.7.2.15 but is evidence for 2.7.2.1'\n",
    "\n",
    "#Line 32 of LJD\n",
    "model.reactions.rxn00166.notes['NOTES'] = 'LJD 2/4/20 - Single carbon source utilization is incorrect due to regulation. Pseudomonas can convert D-serine to pyruvate, but presence of serine does not trigger expression of the gene to do so.'\n",
    "model.reactions.rxn00166.references = {'REFERENCES':'https://link.springer.com/article/10.1007/s00284-016-1021-0'}\n",
    "\n",
    "#Line 38 of LJD\n",
    "reaction138 = Reaction('rxn07430')\n",
    "reaction138.name = '3-methyl-2-oxobutanoate dehydrogenase'\n",
    "reaction138.subsystem = 'Amino Acid Degradation'\n",
    "reaction138.lower_bound = 0\n",
    "reaction138.upper_bound = 1000\n",
    "reaction138.annotation = {'seed.reaction':'rxn07430','ec-code':'1.2.4.4','sbo': 'SBO:0000176','KEGG Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction138.gene_reaction_rule = 'PA14_35520 and PA14_35530'\n",
    "reaction138.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionalilty from SEED'}\n",
    "reaction138.add_metabolites({\n",
    "    model.metabolites.cpd00056_c: -1.0,\n",
    "    model.metabolites.cpd00067_c: -1.0,\n",
    "    model.metabolites.cpd00123_c: -1.0,\n",
    "    model.metabolites.cpd00011_c: 1.0,\n",
    "    model.metabolites.cpd14700_c: 1.0,})\n",
    "model.add_reactions([reaction138])\n",
    "\n",
    "#Line 39 of LJD\n",
    "reaction139 = Reaction('rxn07431')\n",
    "reaction139.name = '2-Methyl-1-hydroxypropyl-ThPP transforming'\n",
    "reaction139.subsystem = 'Amino Acid Degradation'\n",
    "reaction139.lower_bound = -1000\n",
    "reaction139.upper_bound = 1000\n",
    "reaction139.annotation = {'seed.reaction':'rxn07431','ec-code':'1.2.4.4','sbo': 'SBO:0000176','Kegg Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction139.gene_reaction_rule = 'PA14_35520 and PA14_35530'\n",
    "reaction139.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionalilty from SEED'}\n",
    "reaction139.add_metabolites({\n",
    "    model.metabolites.cpd00213_c: -1.0,\n",
    "    model.metabolites.cpd14700_c: -1.0,\n",
    "    model.metabolites.cpd00056_c: 1.0,\n",
    "    model.metabolites.cpd02700_c: 1.0,})\n",
    "model.add_reactions([reaction139])\n",
    "\n",
    "#Line 40 of LJD\n",
    "reaction140 = Reaction('rxn01925')\n",
    "reaction140.name = '2-methylpropanoyl-CoA:enzyme N6-(dihydrolipoyl)lysine S-(2-methylpropanoyl)transferase'\n",
    "reaction140.subsystem = 'Amino Acid Degradation'\n",
    "reaction140.lower_bound = -1000\n",
    "reaction140.upper_bound = 1000\n",
    "reaction140.annotation = {'seed.reaction':'rxn01925','ec-code':'2.3.1.168','sbo': 'SBO:0000176','KEGG Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction140.gene_reaction_rule = 'PA14_35500'\n",
    "reaction140.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionalilty from SEED'}\n",
    "reaction140.add_metabolites({\n",
    "    model.metabolites.cpd00449_c: -1.0,\n",
    "    model.metabolites.cpd00481_c: -1.0,\n",
    "    model.metabolites.cpd00010_c: 1.0,\n",
    "    model.metabolites.cpd02700_c: 1.0,})\n",
    "model.add_reactions([reaction140])\n",
    "\n",
    "#Line 41 of LJD\n",
    "reaction141 = Reaction('rxn01924')\n",
    "reaction141.name = 'isobutyryl-CoA dehydrogenase'\n",
    "reaction141.subsystem = 'Amino Acid Degradation'\n",
    "reaction141.lower_bound = -1000\n",
    "reaction141.upper_bound = 1000\n",
    "reaction141.annotation = {'seed.reaction':'rxn01924','ec-code':'1.3.8.1, 1.3.8.7','sbo': 'SBO:0000176','KEGG Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction141.gene_reaction_rule = 'PA14_66350 or PA14_31580 or PA14_35970 or PA14_44590 or PA14_49080 or PA14_51120 or PA14_52900 or PA14_66040'\n",
    "reaction141.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionality from SEED; metacyc had irreversible forward'}\n",
    "reaction141.add_metabolites({\n",
    "    model.metabolites.cpd00015_c: -1.0,\n",
    "    model.metabolites.cpd00481_c: -1.0,\n",
    "    model.metabolites.cpd00982_c: 1.0,\n",
    "    model.metabolites.cpd02187_c: 1.0,})\n",
    "model.add_reactions([reaction141])\n",
    "\n",
    "#Line 42 of LJD\n",
    "reaction142 = Reaction('rxn02949')\n",
    "reaction142.name = '(S)-3-Hydroxyisobutyryl-CoA hydro-lyase'\n",
    "reaction142.subsystem = 'Amino Acid Degradation'\n",
    "reaction142.lower_bound = -1000\n",
    "reaction142.upper_bound = 1000\n",
    "reaction142.annotation = {'seed.reaction':'rxn02949','ec-code':'4.2.1.17','sbo': 'SBO:0000176','KEGG Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction142.gene_reaction_rule = 'PA14_19740 or PA14_25080 or PA14_40980 or PA14_41950 or PA14_42080'\n",
    "reaction142.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionalilty from SEED; GPR from Kegg'}\n",
    "reaction142.add_metabolites({\n",
    "    model.metabolites.cpd00001_c: -1.0,\n",
    "    model.metabolites.cpd02187_c: -1.0,\n",
    "    model.metabolites.cpd03572_c: 1.0,})\n",
    "model.add_reactions([reaction142])\n",
    "\n",
    "#Line 43 of LJD\n",
    "reaction143 = Reaction('rxn02925')\n",
    "reaction143.name = 'L-Alanine:3-oxopropanoate aminotransferase'\n",
    "reaction143.subsystem = 'Amino Acid Degradation'\n",
    "reaction143.lower_bound = -1000\n",
    "reaction143.upper_bound = 1000\n",
    "reaction143.annotation = {'seed.reaction':'rxn02925','ec-code':'2.6.1.18','sbo': 'SBO:0000176','KEGG Subsystem':'Valine, leucine and isoleucine degradation'}\n",
    "reaction143.gene_reaction_rule = 'PA14_01620 or PA14_70160'\n",
    "reaction143.notes = {'NOTES':'LJD 2/6/20 - Added based on Kegg and Pseudomonas db evidence for partial valine degradation; directionality from SEED'}\n",
    "reaction143.add_metabolites({\n",
    "    model.metabolites.cpd00035_c: -1.0,\n",
    "    model.metabolites.cpd00287_c: -1.0,\n",
    "    model.metabolites.cpd00020_c: 1.0,\n",
    "    model.metabolites.cpd02096_c: 1.0,})\n",
    "model.add_reactions([reaction143])\n",
    "\n",
    "#Line 44 of LJD - flip direction of reaction\n",
    "model.reactions.rxn00868.notes['NOTES'] = 'LJD 2/6/20 - Flipped reaction direction to match SEED and MetaCyc'\n",
    "model.reactions.rxn00868.add_metabolites({\n",
    "    model.metabolites.cpd00004_c: -2.0,\n",
    "    model.metabolites.cpd00067_c: -2.0,\n",
    "    model.metabolites.cpd00650_c: -2.0,\n",
    "    model.metabolites.cpd00003_c: 2.0,\n",
    "    model.metabolites.cpd00120_c: 2.0,})\n",
    "\n",
    "#Remove reactions\n",
    "#Line 31 of LJD\n",
    "model.reactions.rxn00160.remove_from_model()\n",
    "\n",
    "#LJD gene essentiality line 2 - Remove GPR \n",
    "model.reactions.rxn01355.gene_reaction_rule = ''\n",
    "\n",
    "#LJD other changes line 2 - change Ec-code\n",
    "model.reactions.rxn05155.annotation[\"ec-code\"]=\"7.4.2.1\"\n",
    "\n",
    "#_____________________________________________________________________________________________________\n",
    "#Add SBO numbers to genes per memote report\n",
    "model.genes.PA14_17050.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_28590.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_47840.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_47860.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_48000.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_36050.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_48010.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_31530.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_13090.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_63250.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_71140.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_71420.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_71410.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_71280.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_71260.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_10610.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_39100.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_10650.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_10640.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_10590.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_10570.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_35520.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_35500.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_35970.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_44590.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_51120.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_66040.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_49080.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_19740.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_40980.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_70160.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_48440.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_51990.annotation['sbo'] = 'SBO:0000243'\n",
    "model.genes.PA14_37340.annotation['sbo'] = 'SBO:0000243'\n",
    "\n",
    "#Add kegg IDs to genes\n",
    "model.genes.PA14_48440.annotation['kegg.genes'] = 'pau:PA14_48440'\n",
    "model.genes.PA14_51990.annotation['kegg.genes'] = 'pau:PA14_51990'\n",
    "model.genes.PA14_37340.annotation['kegg.genes'] = 'pau:PA14_37340'\n",
    "model.genes.PA14_28590.annotation['kegg.genes'] = 'pau:PA14_28590'\n",
    "model.genes.PA14_17050.annotation['kegg.genes'] = 'pau:PA14_17050'\n",
    "model.genes.PA14_47840.annotation['kegg.genes'] = 'pau:PA14_47840'\n",
    "model.genes.PA14_47860.annotation['kegg.genes'] = 'pau:PA14_47860'\n",
    "model.genes.PA14_48000.annotation['kegg.genes'] = 'pau:PA14_48000'\n",
    "model.genes.PA14_48010.annotation['kegg.genes'] = 'pau:PA14_48010'\n",
    "model.genes.PA14_36050.annotation['kegg.genes'] = 'pau:PA14_36050'\n",
    "model.genes.PA14_13090.annotation['kegg.genes'] = 'pau:PA14_13090'\n",
    "model.genes.PA14_63250.annotation['kegg.genes'] = 'pau:PA14_63250'\n",
    "model.genes.PA14_31530.annotation['kegg.genes'] = 'pau:PA14_31530'\n",
    "model.genes.PA14_71140.annotation['kegg.genes'] = 'pau:PA14_71140'\n",
    "model.genes.PA14_71410.annotation['kegg.genes'] = 'pau:PA14_71410'\n",
    "model.genes.PA14_71420.annotation['kegg.genes'] = 'pau:PA14_71420'\n",
    "model.genes.PA14_71260.annotation['kegg.genes'] = 'pau:PA14_71260'\n",
    "model.genes.PA14_71280.annotation['kegg.genes'] = 'pau:PA14_71280'\n",
    "model.genes.PA14_10610.annotation['kegg.genes'] = 'pau:PA14_10610'\n",
    "model.genes.PA14_39100.annotation['kegg.genes'] = 'pau:PA14_39100'\n",
    "model.genes.PA14_10650.annotation['kegg.genes'] = 'pau:PA14_10650'\n",
    "model.genes.PA14_10640.annotation['kegg.genes'] = 'pau:PA14_10640'\n",
    "model.genes.PA14_10590.annotation['kegg.genes'] = 'pau:PA14_10590'\n",
    "model.genes.PA14_10570.annotation['kegg.genes'] = 'pau:PA14_10570'\n",
    "model.genes.PA14_35520.annotation['kegg.genes'] = 'pau:PA14_35520'\n",
    "model.genes.PA14_35500.annotation['kegg.genes'] = 'pau:PA14_35500'\n",
    "model.genes.PA14_66040.annotation['kegg.genes'] = 'pau:PA14_66040'\n",
    "model.genes.PA14_35970.annotation['kegg.genes'] = 'pau:PA14_35970'\n",
    "model.genes.PA14_51120.annotation['kegg.genes'] = 'pau:PA14_51120'\n",
    "model.genes.PA14_49080.annotation['kegg.genes'] = 'pau:PA14_49080'\n",
    "model.genes.PA14_44590.annotation['kegg.genes'] = 'pau:PA14_44590'\n",
    "model.genes.PA14_40980.annotation['kegg.genes'] = 'pau:PA14_40980'\n",
    "model.genes.PA14_19740.annotation['kegg.genes'] = 'pau:PA14_19740'\n",
    "model.genes.PA14_70160.annotation['kegg.genes'] = 'pau:PA14_70160'\n",
    "\n",
    "#Correct sbo number on reaction\n",
    "model.reactions.ATPM.annotation['sbo'] = 'SBO:0000176'\n",
    "\n",
    "#Delete mislabeled metabolite annotation\n",
    "del model.metabolites.cJB00125_c.annotation['seed.compound']\n",
    "\n",
    "#Delete incorrect seed annotations on rPY and rJB reactions\n",
    "for x in model.reactions:\n",
    "    if x.id[0:3] == \"rPY\":\n",
    "        del x.annotation['seed.reaction']\n",
    "    elif x.id[0:3] == \"rJB\":\n",
    "        del x.annotation['seed.reaction']\n",
    "        \n",
    "#_____________________________________________________________________________________________________\n",
    "#Reaction Imbalance Fixes\n",
    "\n",
    "#Fix rxn02749 imbalance\n",
    "model.metabolites.cpd01700_c.formula = 'C5H6O5'\n",
    "\n",
    "#Fix rxn13839 imbalance\n",
    "model.metabolites.cpd03913_c.formula = 'C45H60N6O12'\n",
    "model.metabolites.cpd03913_c.charge = -2\n",
    "model.reactions.rxn13839.add_metabolites({model.metabolites.cpd00067_c: 5.0})\n",
    "model.metabolites.cpd03914_c.formula = 'C45H56CoN6O12'\n",
    "model.metabolites.cpd03914_c.charge = -4\n",
    "\n",
    "#Makes rxn11544 unbalanced, fix\n",
    "model.metabolites.cpd03832_c.formula = 'C45H56N4O14'\n",
    "model.metabolites.cpd03832_c.charge = -4\n",
    "\n",
    "#Makes rxn03512 unbalanced, fix\n",
    "model.reactions.rxn03512.add_metabolites({model.metabolites.cpd00067_c: -2.0})\n",
    "\n",
    "#Fix rxn00105 imbalance\n",
    "model.metabolites.cpd00355_c.formula = 'C11H14N2O8P'\n",
    "model.metabolites.cpd00355_c.charge = -1\n",
    "#_____________________________________________________________________________________________________\n",
    "#Correct AR's dates on notes on rxn & metabolites modified\n",
    "model.reactions.rxn05964.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn06079.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn06087.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn05901.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn06140.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn06043.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn04684.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn00538.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn05829.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn05824.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn00182.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn02749.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn02751.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn00735.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn03062.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn01208.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn03436.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn03435.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn02187.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn03068.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn03437.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn00804.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn01573.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.reactions.rxn00904.notes['NOTES'] = '*AR 11/28/18'\n",
    "model.metabolites.cpd00457_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd00285_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd12185_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd11611_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd00049_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd01700_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd03593_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd02605_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd10162_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "model.metabolites.cpd02569_c.notes = {'NOTES':'*AR 11/28/18'}\n",
    "#_____________________________________________________________________________________________________\n",
    "#Add peptidoglycan polymer n-1 subunits metabolite to model, balance associated reactions, add synthesis reaction\n",
    "\n",
    "#Add n-1 subunits peptidoglycan polymer to model\n",
    "cpd15666= Metabolite('cpd15666_c',\n",
    "                        formula='C40H63N8O21R',\n",
    "                        name='Peptidoglycan polymer (n-1 subunits)',\n",
    "                        compartment='cytosol',\n",
    "                        charge = -2)\n",
    "model.add_metabolites([cpd15666])\n",
    "model.metabolites.cpd15666_c.notes = '*DDP 5/13/20 added'\n",
    "model.metabolites.cpd15666_c.annotation = {'sbo': 'SBO:0000247','seed.compound':'cpd15666'}\n",
    "\n",
    "#Correct rxn10194 imbalance\n",
    "model.reactions.rxn10194.add_metabolites({model.metabolites.cpd15666_c: -1,\n",
    "                                          model.metabolites.cpd15665_c: 1})\n",
    "\n",
    "#Correct rxn10195 imbalance\n",
    "model.reactions.rxn10195.add_metabolites({model.metabolites.cpd15666_c: -1,\n",
    "                                          model.metabolites.cpd15665_c: 1})\n",
    "\n",
    "#Correct rxn10196 imbalance\n",
    "model.reactions.rxn10196.add_metabolites({model.metabolites.cpd15666_c: -1,\n",
    "                                          model.metabolites.cpd15665_c: 1})\n",
    "\n",
    "#Correct rxn10199 imbalance by adding n-1 subunits\n",
    "model.reactions.rxn10199.add_metabolites({model.metabolites.cpd15666_c: -1})\n",
    "\n",
    "#Add rxn12751 for n-1 subunits synthesis\n",
    "rxn12751 = Reaction('rxn12751')\n",
    "rxn12751.name = 'Peptidoglycan subunit synthesis P.putida KT2440'\n",
    "rxn12751.lower_bound = -1000\n",
    "rxn12751.upper_bound = 1000\n",
    "rxn12751.annotation = {'sbo': 'SBO:0000657','seed.reaction':'rxn12751','iJN746':'PPTGS_KT'}\n",
    "rxn12751.notes = '*DDP 5/19/20 - added for peptidoglycan biosynthesis'\n",
    "rxn12751.references = {'REFERENCES':'Nogales, J., Palsson, B. Ø., & Thiele, I. (2008). A genome-scale metabolic reconstruction of Pseudomonas putida KT2440: i JN746 as a cell factory. BMC systems biology, 2(1), 79.'}\n",
    "rxn12751.add_metabolites({\n",
    "    model.metabolites.cpd03495_c: -1.0,\n",
    "    model.metabolites.cpd02229_c: 1.0,\n",
    "    model.metabolites.cpd15666_c: 1.0,})\n",
    "model.add_reactions([rxn12751])\n",
    "#_____________________________________________________________________________________________________\n",
    "#Add periplasm compartment for ATP synthase loop fixing\n",
    "\n",
    "#Add periplasm hydrogen\n",
    "cpd00067_p= Metabolite('cpd00067_p',\n",
    "                        formula='H',\n",
    "                        name='H+',\n",
    "                        compartment='periplasm',\n",
    "                        charge = 1)\n",
    "model.add_metabolites([cpd00067_p])\n",
    "model.metabolites.cpd00067_p.notes = '*DDP 5/19/20 added to periplasm compartment to fix ATP synthase loop'\n",
    "model.metabolites.cpd00067_p.annotation = {'sbo': 'SBO:0000247','bigg.metabolite': 'h',\n",
    " 'biocyc': 'META:PROTON',\n",
    " 'chebi': ['CHEBI:10744',\n",
    "  'CHEBI:13357',\n",
    "  'CHEBI:15378',\n",
    "  'CHEBI:24636',\n",
    "  'CHEBI:5584'],\n",
    " 'hmdb': 'HMDB59597',\n",
    " 'kegg.compound': 'C00080',\n",
    " 'metanetx.chemical': 'MNXM1',\n",
    " 'seed.compound':'cpd00067',\n",
    " 'unipathway.compound': 'UPC00080'}\n",
    "\n",
    "#rxn10042 move H+ to periplasm\n",
    "model.reactions.rxn10042.add_metabolites({\n",
    "    model.metabolites.cpd00067_p: -4.0,\n",
    "    model.metabolites.cpd00067_e: 4.0})\n",
    "model.reactions.rxn10042.notes['NOTES'] = '*DDP 05/21/20 - moved extracellular hydrogen to periplasm to resolve ATP synthase loop; *JAB 02/19/14 - removed PA1104 or PA1697 isozymes from ATP synthase operon' \n",
    "\n",
    "\n",
    "#rJB00264 move H+ to periplasm\n",
    "model.reactions.rJB00264.add_metabolites({\n",
    "    model.metabolites.cpd00067_p: 4.0,\n",
    "    model.metabolites.cpd00067_e: -4.0})\n",
    "model.reactions.rJB00264.notes['NOTES'] = '*DDP 05/21/20 - moved extracellular hydrogen to periplasm to resolve ATP synthase loop; *PY 06/11/14 Changed leqrightarrow to rightarrow *JAB 04/16/14 - ub8gequb9 fix; JAB 01/31/14 - switched reaction directionality to match MetaCyc'\n",
    "\n",
    "#rxn13726 move H+ to periplasm\n",
    "model.reactions.rxn13726.add_metabolites({\n",
    "    model.metabolites.cpd00067_p: 4.0,\n",
    "    model.metabolites.cpd00067_e: -4.0})\n",
    "model.reactions.rxn13726.notes['NOTES'] = '*DDP 05/21/20 - moved extracellular hydrogen to periplasm to resolve ATP synthase loop; *JAB 04/16/14 - fixed ub8 ref; JAB 01/31/14 - switched reaction directionality to match MetaCyc'\n",
    "\n",
    "#rxn13689 move H+ to periplasm\n",
    "model.reactions.rxn13689.add_metabolites({\n",
    "    model.metabolites.cpd00067_p: 4.0,\n",
    "    model.metabolites.cpd00067_e: -4.0})\n",
    "model.reactions.rxn13689.notes['NOTES'] = '*DDP 05/21/20 - moved extracellular hydrogen to periplasm to resolve ATP synthase loop; *JAB 04/16/14 - fixed ub8 ref;'\n",
    "\n",
    "#rJB00271 move H+ to periplasm\n",
    "model.reactions.rJB00271.add_metabolites({\n",
    "    model.metabolites.cpd00067_p: 4.0,\n",
    "    model.metabolites.cpd00067_e: -4.0})\n",
    "model.reactions.rJB00271.notes['NOTES'] = '*DDP 05/21/20 - moved extracellular hydrogen to periplasm to resolve ATP synthase loop; *JAB 04/16/14 - ub8gequb9 fix;'\n",
    "\n",
    "#________________________________________________________\n",
    "#Corrections to reaction to resolve ATP generating loop\n",
    "\n",
    "#Correct reversiibility of rxn00251\n",
    "model.reactions.rxn00251.lower_bound = 0\n",
    "model.reactions.rxn00251.notes = {'NOTES':'*DDP 05/28/20 changed to irreversible'}\n",
    "model.reactions.rxn00251.references = 'REFERENCES: Kai, Y., Matsumura, H., & Izui, K. (2003). Phosphoenolpyruvate carboxylase: three-dimensional structure and molecular mechanisms. Archives of Biochemistry and Biophysics, 414(2), 170-179.'\n",
    "\n",
    "#Correct reversiibility of rxn13919\n",
    "model.reactions.rxn13919.lower_bound = 0\n",
    "model.reactions.rxn13919.notes = {'NOTES':'*DDP 05/28/20 changed to irreversible since efflux transporter  *JAB 12/04/14 - added second cation efflux transporter from PA14'}\n",
    "\n",
    "#Correct reversibility of rxn05145\n",
    "model.reactions.rxn05145.lower_bound = 0\n",
    "model.reactions.rxn05145.notes = {'NOTES':'*DDP 05/28/20 changed to irreversible'}\n",
    "model.reactions.rxn05145.references = 'REFERENCES: Locher, K. P. (2016). Mechanistic diversity in ATP-binding cassette (ABC) transporters. Nature structural & molecular biology, 23(6), 487.'\n",
    "\n",
    "#Correct reversibility of rxn05158\n",
    "model.reactions.rxn05158.lower_bound = 0\n",
    "model.reactions.rxn05145.notes = {'NOTES':'*DDP 05/28/20 changed to irreversible'}\n",
    "model.reactions.rxn05145.references = 'REFERENCES: Schneider, E. (2001). ABC transporters catalyzing carbohydrate uptake. Research in microbiology, 152(3-4), 303-310.'\n",
    "\n",
    "#Correct rxn01452 GPR\n",
    "model.reactions.rxn01452.gene_reaction_rule = 'PA14_17860 or PA14_43460'\n",
    "model.reactions.rxn01452.notes['NOTES'] = '*DDP 06/09/20 updated GPR, previously was wrong gene (PA14_53950) *JAB 12/19/13 from iBsu1103: Bsu2414, Bsu gene mappings are wrong (peg to locus tag is offset somewhere, but gene identifiers and start sequence location seems to match)'\n",
    "\n",
    "#Change rxn00097 to irreversible\n",
    "model.reactions.rxn00097.lower_bound = 0\n",
    "model.reactions.rxn00097.notes = {'NOTES':'*DDP 06/16/20 changed to irreversible to resolve ATP generating loop'}\n",
    "\n",
    "#Change rxn00161 to irreversible\n",
    "model.reactions.rxn00161.lower_bound = 0\n",
    "model.reactions.rxn00161.notes['NOTES'] = '*DDP 06/16/20 changed to irreversible to resolve ATP generating loop  *JAB 06/26/14 - experimental evidence of NADP cofactor utilization; *JCT 02/17/02 - changed to reversible'\n",
    "\n",
    "#Change rxn08066 to irreversible\n",
    "model.reactions.rxn08066.lower_bound = 0\n",
    "model.reactions.rxn08066.notes['NOTES'] = '*DDP 06/16/20 changed to irreversible to resolve ATP generating loop  *ASB 03/28/14 - tentative subsystem assignment'\n",
    "\n",
    "#Change rxn13842 to irreversible\n",
    "model.reactions.rxn13842.lower_bound = 0\n",
    "model.reactions.rxn13842.notes['NOTES'] = '*DDP 06/16/20 changed to irreversible to resolve ATP generating loop  *ASB 03/28/14 - tentative subsystem assignment'\n",
    "#_____________________________________________________________________________________________________\n",
    "#Carbon source prediction fixes\n",
    "\n",
    "#Fix L-valine and acetoacetate predictions - make rxn00178 irreversible\n",
    "model.reactions.rxn00178.lower_bound = 0\n",
    "model.reactions.rxn00178.notes['NOTES'] = '*DDP 6/23/20 Making irreversible to fix L-valine and acetoacetate growth predictions *LJD 1/21/20 - Adding this reaction allows growth on L-lysine. For now including all six genes associated with this enzyme in an or statement for lack of more information '\n",
    "\n",
    "\n",
    "#Add PA14_68390 to model & update rxn00708 GPR - doesn't fix prediction\n",
    "model.reactions.rxn00708.gene_reaction_rule = 'PA14_17450 or PA14_68390'\n",
    "model.genes.PA14_68390.name = 'putative hydrolase'\n",
    "model.genes.PA14_68390.annotation = {'kegg.genes':'pau:PA14_68390','sbo':'SBO:0000243'}\n",
    "#_____________________________________________________________________________________________________\n",
    "#Add in all carbon source testing extracellular metabolites & exchanges\n",
    "\n",
    "#Add metabolites\n",
    "cpd01949 = Metabolite('cpd01949_e',\n",
    "                 formula = 'C4H10O2',\n",
    "                 name = '(SS)-23-Butanediol',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd01949])\n",
    "\n",
    "cpd00361 = Metabolite('cpd00361_e',\n",
    "                 formula = 'C4H8O2',\n",
    "                 name = 'ACTN',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd00361])\n",
    "\n",
    "cpd00158 = Metabolite('cpd00158_e',\n",
    "                 formula = 'C12H22O11',\n",
    "                 name = 'CELB',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd00158])\n",
    "\n",
    "cpd00155 = Metabolite('cpd00155_e',\n",
    "                 formula = 'C6H9O5',\n",
    "                 name = 'Glycogen',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -1)\n",
    "model.add_metabolites([cpd00155])\n",
    "\n",
    "cpd00040 = Metabolite('cpd00040_e',\n",
    "                 formula = 'C2HO3',\n",
    "                 name = 'Glyoxalate',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -1)\n",
    "model.add_metabolites([cpd00040])\n",
    "\n",
    "cpd00294 = Metabolite('cpd00294_e',\n",
    "                 formula = 'C10H13N5O6P',\n",
    "                 name = 'dAMP',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -1)\n",
    "model.add_metabolites([cpd00294])\n",
    "\n",
    "cpd03320 = Metabolite('cpd03320_e',\n",
    "                 formula = 'C8H8O3',\n",
    "                 name = '3-Hydroxyphenylacetate',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd03320])\n",
    "\n",
    "cpd00094 = Metabolite('cpd00094_e',\n",
    "                 formula = 'C4H5O3',\n",
    "                 name = '2-Oxobutyrate',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -1)\n",
    "model.add_metabolites([cpd00094])\n",
    "\n",
    "cpd00227 = Metabolite('cpd00227_e',\n",
    "                 formula = 'C4H9NO3',\n",
    "                 name = 'L-Homoserine',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd00227])\n",
    "\n",
    "cpd00599 = Metabolite('cpd00599_e',\n",
    "                 formula = 'C7H5O3',\n",
    "                 name = 'salicylate',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = -1)\n",
    "model.add_metabolites([cpd00599])\n",
    "\n",
    "cpd00361 = Metabolite('cpd00361_e',\n",
    "                 formula = 'C4H8O2',\n",
    "                 name = 'ACTN',\n",
    "                 compartment = 'extracellular',\n",
    "                 charge = 0)\n",
    "model.add_metabolites([cpd00361])\n",
    "\n",
    "\n",
    "#add exchange rxns\n",
    "cpd01949 = Reaction('EX_cpd01949_e')\n",
    "cpd01949.name = 'cpd01949 exchange'\n",
    "cpd01949.lower_bound = 0\n",
    "cpd01949.upper_bound = 1000\n",
    "cpd01949.add_metabolites({\n",
    "    model.metabolites.cpd01949_e: -1.0})\n",
    "model.add_reactions([cpd01949])\n",
    " \n",
    "cpd00361 = Reaction('EX_cpd00361_e')\n",
    "cpd00361.name = 'cpd00361 exchange'\n",
    "cpd00361.lower_bound = 0\n",
    "cpd00361.upper_bound = 1000\n",
    "cpd00361.add_metabolites({\n",
    "    model.metabolites.cpd00361_e: -1.0})\n",
    "model.add_reactions([cpd00361])\n",
    "\n",
    "cpd00158 = Reaction('EX_cpd00158_e')\n",
    "cpd00158.name = 'cpd00158 exchange'\n",
    "cpd00158.lower_bound = 0\n",
    "cpd00158.upper_bound = 1000\n",
    "cpd00158.add_metabolites({\n",
    "    model.metabolites.cpd00158_e: -1.0})\n",
    "model.add_reactions([cpd00158])\n",
    "\n",
    "cpd00155 = Reaction('EX_cpd00155_e')\n",
    "cpd00155.name = 'cpd00155 exchange'\n",
    "cpd00155.lower_bound = 0\n",
    "cpd00155.upper_bound = 1000\n",
    "cpd00155.add_metabolites({\n",
    "    model.metabolites.cpd00155_e: -1.0})\n",
    "model.add_reactions([cpd00155])\n",
    " \n",
    "cpd00040 = Reaction('EX_cpd00040_e')\n",
    "cpd00040.name = 'cpd00040 exchange'\n",
    "cpd00040.lower_bound = 0\n",
    "cpd00040.upper_bound = 1000\n",
    "cpd00040.add_metabolites({\n",
    "    model.metabolites.cpd00040_e: -1.0})\n",
    "model.add_reactions([cpd00040])\n",
    " \n",
    "cpd00294 = Reaction('EX_cpd00294_e')\n",
    "cpd00294.name = 'cpd00294 exchange'\n",
    "cpd00294.lower_bound = -1000\n",
    "cpd00294.upper_bound = 1000\n",
    "cpd00294.add_metabolites({\n",
    "    model.metabolites.cpd00294_e: -1.0})\n",
    "model.add_reactions([cpd00294])\n",
    "\n",
    "cpd03320 = Reaction('EX_cpd03320_e')\n",
    "cpd03320.name = 'cpd03320 exchange'\n",
    "cpd03320.lower_bound = 0\n",
    "cpd03320.upper_bound = 1000\n",
    "cpd03320.add_metabolites({\n",
    "    model.metabolites.cpd03320_e: -1.0})\n",
    "model.add_reactions([cpd03320])\n",
    "\n",
    "cpd00094 = Reaction('EX_cpd00094_e')\n",
    "cpd00094.name = 'cpd00094 exchange'\n",
    "cpd00094.lower_bound = 0\n",
    "cpd00094.upper_bound = 1000\n",
    "cpd00094.add_metabolites({\n",
    "    model.metabolites.cpd00094_e: -1.0})\n",
    "model.add_reactions([cpd00094])\n",
    "\n",
    "cpd00227 = Reaction('EX_cpd00227_e')\n",
    "cpd00227.name = 'cpd00227 exchange'\n",
    "cpd00227.lower_bound = -1000\n",
    "cpd00227.upper_bound = 1000\n",
    "cpd00227.add_metabolites({\n",
    "    model.metabolites.cpd00227_e: -1.0})\n",
    "model.add_reactions([cpd00227])\n",
    "\n",
    "cpd00599 = Reaction('EX_cpd00599_e')\n",
    "cpd00599.name = 'cpd00599 exchange'\n",
    "cpd00599.lower_bound = 0\n",
    "cpd00599.upper_bound = 1000\n",
    "cpd00599.add_metabolites({\n",
    "    model.metabolites.cpd00599_e: -1.0})\n",
    "model.add_reactions([cpd00599])\n",
    " \n",
    "#cpd00361 = Reaction('EX_cpd00361_e')\n",
    "#cpd00361.name = 'cpd00361 exchange'\n",
    "#cpd00361.lower_bound = 0\n",
    "#cpd00361.upper_bound = 1000\n",
    "#cpd00361.add_metabolites({\n",
    "#    model.metabolites.cpd00361_e: -1.0})\n",
    "#model.add_reactions([cpd00361]) \n",
    "#_____________________________________________________________________________________________________\n",
    "model.metabolites.cpd15666_c.notes = {'NOTES':'DDP 5/13/20 added'}\n",
    "model.metabolites.cpd00067_p.notes = {'NOTES':'*DDP 5/19/20 added to periplasm compartment to fix ATP synthase loop'}\n",
    "model.reactions.rxn12751.notes = {'NOTES':''}\n",
    "\n",
    "#Add rxn00994\n",
    "rxn00994 = Reaction('rxn00994')\n",
    "rxn00994.name = 'Butanoyl-CoA:acetoacetate CoA-transferase'\n",
    "rxn00994.lower_bound = -1000\n",
    "rxn00994.upper_bound = 1000\n",
    "rxn00994.annotation = {'ec-code':'2.8.3.8'}\n",
    "rxn00994.gene_reaction_rule = ''\n",
    "rxn00994.notes['NOTES'] = '*DDP 8/11/20 BLASTn showed significant evidence, added as part of butryate metabolism. No gene name but position 3446881 to 3447254.'\n",
    "rxn00994.references = ''\n",
    "rxn00994.add_metabolites({\n",
    "    model.metabolites.cpd00120_c: -1.0,\n",
    "    model.metabolites.cpd00142_c: -1.0,\n",
    "    model.metabolites.cpd00211_c: 1.0,\n",
    "    model.metabolites.cpd00279_c: 1.0})\n",
    "model.add_reactions([rxn00994])\n",
    "\n",
    "\n",
    "#Add rxn00875\n",
    "rxn00875 = Reaction('rxn00875')\n",
    "rxn00875.name = 'Butanoyl-CoA:acetate CoA-transferase'\n",
    "rxn00875.lower_bound = -1000\n",
    "rxn00875.upper_bound = 1000\n",
    "rxn00875.annotation = {'ec-code':'2.8.3.8'}\n",
    "rxn00875.gene_reaction_rule = ''\n",
    "rxn00875.notes['NOTES'] = '*DDP 8/11/20 BLASTn showed significant evidence, added as part of butryate metabolism. No gene name but position 3446881 to 3447254.'\n",
    "rxn00875.references = ''\n",
    "rxn00875.add_metabolites({\n",
    "    model.metabolites.cpd00029_c: -1.0,\n",
    "    model.metabolites.cpd00120_c: -1.0,\n",
    "    model.metabolites.cpd00022_c: 1.0,\n",
    "    model.metabolites.cpd00211_c: 1.0})\n",
    "model.add_reactions([rxn00875])\n",
    "\n",
    "#Remove SPONTANEOUS, unassigned, and Unassigned GPR from associated rxns\n",
    "model.reactions.rxn13912.gene_reaction_rule = ''\n",
    "model.reactions.rxn13912.notes['NOTES'] = '*DDP 8/12/20 removed unassigned GPR *JAB 01/19/15 - no good hit to PA3155, leaving as unassigned to complete pathway, but slim chance there is link to PA14_69150 *ASB 03/24/14 - tentative subsystem assignment'\n",
    "model.reactions.rxn05319.gene_reaction_rule = ''\n",
    "model.reactions.rxn05319.notes['NOTES'] = '*DDP 8/12/20 removed SPONTANEOUS GPR *ASB 12/08/14 - changed gene name'\n",
    "model.reactions.rxn05468.gene_reaction_rule = ''\n",
    "model.reactions.rxn05468.notes['NOTES']= '*DDP 8/12/20 removed SPONTANEOUS GPR *ASB 12/08/14 - changed gene name'\n",
    "model.reactions.rxn05467.gene_reaction_rule = ''\n",
    "model.reactions.rxn05467.notes['NOTES'] = '*DDP 8/12/20 removed SPONTANEOUS GPR *ASB 12/08/14 - changed gene name'\n",
    "model.reactions.rxn05064.gene_reaction_rule = ''\n",
    "model.reactions.rxn05064.notes['NOTES'] = '*DDP 8/12/20 removed SPONTANEOUS GPR *ASB 12/08/14 - changed gene name'\n",
    "model.reactions.rxn10577.gene_reaction_rule = ''\n",
    "model.reactions.rxn10577.notes['NOTES'] = '*DDP 8/12/20 removed SPONTANEOUS GPR *ASB 12/08/14 - changed gene name'\n",
    "model.reactions.rJB00285.gene_reaction_rule = ''\n",
    "model.reactions.rJB00285.notes['NOTES'] = '*DDP 8/12/20 removed Unassigned GPR *JAB 01/18/15 - cis-2-decenoic acid is DSF (diffusible signal factor)'\n",
    "\n",
    "#Delete SPONTANEOUS, unassigned, and Unassigned genes from model\n",
    "model.genes.remove('Unassigned')\n",
    "model.genes.remove('SPONTANEOUS')\n",
    "model.genes.remove('unassigned')\n",
    "#_____________________________________________________________________________________________________\n",
    "#BOF work - organize substrates and add LPS and biomass metabolites to BOF\n",
    "\n",
    "#add protein_c metabolite\n",
    "protein_c = Metabolite('protein_c',\n",
    "                       formula = '',\n",
    "                       name = 'Protein',\n",
    "                       compartment = 'cytosol')\n",
    "model.add_metabolites([protein_c])\n",
    "model.metabolites.protein_c.notes = {'NOTES':'*DDP 11/17/20 - added to represent proteins required for growth'}\n",
    "model.metabolites.protein_c.annotation = {'sbo':'SBO:0000247'}\n",
    "#add protein_rxn\n",
    "protein_rxn = Reaction('protein_rxn')\n",
    "protein_rxn.name = 'protein_rxn'\n",
    "protein_rxn.subsystem = 'cytosol'\n",
    "protein_rxn.lower_bound = 0\n",
    "protein_rxn.upper_bound = 1000\n",
    "protein_rxn.annotation = {'sbo': 'SBO:0000632'}\n",
    "protein_rxn.notes = {'NOTES':'*DDP 11/17/20 - added to bin metabolites to protein required for growth'}\n",
    "protein_rxn.add_metabolites({\n",
    "model.metabolites.cpd00084_c: -0.037574,\n",
    "model.metabolites.cpd00065_c: -0.055748,\n",
    "model.metabolites.cpd00060_c: -0.076901,\n",
    "model.metabolites.cpd00119_c: -0.081188,\n",
    "model.metabolites.cpd00069_c: -0.094418,\n",
    "model.metabolites.cpd00132_c: -0.098829,\n",
    "model.metabolites.cpd00039_c: -0.10776,\n",
    "model.metabolites.cpd00066_c: -0.13212,\n",
    "model.metabolites.cpd00322_c: -0.15583,\n",
    "model.metabolites.cpd00161_c: -0.15734,\n",
    "model.metabolites.cpd00053_c: -0.16033,\n",
    "model.metabolites.cpd00129_c: -0.18965,\n",
    "model.metabolites.cpd00041_c: -0.19864,\n",
    "model.metabolites.cpd00054_c: -0.20723,\n",
    "model.metabolites.cpd00023_c: -0.22687,\n",
    "model.metabolites.cpd00156_c: -0.25557,\n",
    "model.metabolites.cpd00051_c: -0.28443,\n",
    "model.metabolites.cpd00033_c: -0.31256,\n",
    "model.metabolites.cpd00035_c: -0.43461,\n",
    "model.metabolites.cpd00107_c: -0.46321,\n",
    "model.metabolites.protein_c: 1})\n",
    "model.add_reactions([protein_rxn])\n",
    "\n",
    "#add rna_c metabolite\n",
    "rna_c = Metabolite('rna_c',\n",
    "                formula = '',\n",
    "                name = 'RNA',\n",
    "                compartment = 'cytosol')\n",
    "model.add_metabolites([rna_c])\n",
    "model.metabolites.rna_c.notes = {'NOTES':'*DDP 11/17/20 - added to represent RNA required for growth'}\n",
    "model.metabolites.rna_c.annotation = {'sbo': 'SBO:0000247'}\n",
    "#add rna_rxn\n",
    "rna_rxn = Reaction('rna_rxn')\n",
    "rna_rxn.name = 'rna_rxn'\n",
    "rna_rxn.subsystem = 'cytosol'\n",
    "rna_rxn.lower_bound = 0\n",
    "rna_rxn.upper_bound = 1000\n",
    "rna_rxn.annotation = {'sbo': 'SBO:0000632'}\n",
    "rna_rxn.notes = {'NOTES':'*DDP 11/17/20 - added to bin metabolites to RNA required for growth'}\n",
    "rna_rxn.add_metabolites({\n",
    "model.metabolites.cpd00091_c: -0.10137,\n",
    "model.metabolites.cpd00018_c: -0.10146,\n",
    "model.metabolites.cpd00126_c: -0.19942,\n",
    "model.metabolites.cpd00046_c: -0.19949,\n",
    "model.metabolites.rna_c: 1})\n",
    "model.add_reactions([rna_rxn])\n",
    "\n",
    "#add dna_c metabolite\n",
    "dna_c = Metabolite('dna_c',\n",
    "                formula = '',\n",
    "                name = 'DNA',\n",
    "                compartment = 'cytosol')\n",
    "model.add_metabolites([dna_c])\n",
    "model.metabolites.dna_c.notes = {'NOTES':'*DDP 11/17/20 - added to represent DNA required for growth'}\n",
    "model.metabolites.dna_c.annotation = {'sbo': 'SBO:0000247'}\n",
    "#add dna_rxn\n",
    "dna_rxn = Reaction('dna_rxn')\n",
    "dna_rxn.name = 'dna_rxn'\n",
    "dna_rxn.subsystem = 'cytosol'\n",
    "dna_rxn.lower_bound = 0\n",
    "dna_rxn.upper_bound = 1000\n",
    "dna_rxn.annotation = {'sbo': 'SBO:0000632'}\n",
    "dna_rxn.notes = {'NOTES':'*DDP 11/17/20 - added to bin metabolites to DNA required for growth'}\n",
    "dna_rxn.add_metabolites({\n",
    "model.metabolites.cpd00294_c: -0.015972,\n",
    "model.metabolites.cpd00298_c: -0.015972,\n",
    "model.metabolites.cpd00206_c: -0.031422,\n",
    "model.metabolites.cpd00296_c: -0.031422,\n",
    "model.metabolites.dna_c: 1})\n",
    "model.add_reactions([dna_rxn])\n",
    "\n",
    "#add lipid_c metabolite\n",
    "lipid_c = Metabolite('lipid_c',\n",
    "                formula = '',\n",
    "                name = 'Lipid',\n",
    "                compartment = 'cytosol')\n",
    "model.add_metabolites([lipid_c])\n",
    "model.metabolites.lipid_c.notes = {'NOTES':'*DDP 11/17/20 - added to represent lipids required for growth'}\n",
    "model.metabolites.lipid_c.annotation = {'sbo': 'SBO:0000247'}\n",
    "#add lipid_rxn\n",
    "lipid_rxn = Reaction('lipid_rxn')\n",
    "lipid_rxn.name = 'lipid_rxn'\n",
    "lipid_rxn.subsystem = 'cytosol'\n",
    "lipid_rxn.lower_bound = 0\n",
    "lipid_rxn.upper_bound = 1000\n",
    "lipid_rxn.annotation = {'sbo': 'SBO:0000632'}\n",
    "lipid_rxn.notes = {'NOTES':'*DDP 11/17/20 - added to bin metabolites to lipid required for growth'}\n",
    "lipid_rxn.add_metabolites({\n",
    "model.metabolites.cpd15430_c: -0.00037815,\n",
    "model.metabolites.cpd15540_c: -0.00042166,\n",
    "model.metabolites.cpd15533_c: -0.0020692,\n",
    "model.metabolites.cpd15539_c: -0.0037982,\n",
    "model.metabolites.cpd15429_c: -0.0041649,\n",
    "model.metabolites.cpd15428_c: -0.0078591,\n",
    "model.metabolites.cpd15541_c: -0.011557,\n",
    "model.metabolites.cpd15431_c: -0.011889,\n",
    "model.metabolites.cpd15538_c: -0.012557,\n",
    "model.metabolites.cpd15532_c: -0.020287,\n",
    "model.metabolites.cpd17066_c: -0.025523,\n",
    "model.metabolites.cpd15531_c: -0.047824,\n",
    "model.metabolites.cpd15534_c: -0.068198,\n",
    "model.metabolites.cPY00138_c: -0.00053681,\n",
    "model.metabolites.cPY00132_c: -0.00079222,\n",
    "model.metabolites.cPY00140_c: -0.00097843,\n",
    "model.metabolites.cPY00124_c: -0.0013602,\n",
    "model.metabolites.cPY00135_c: -0.0015844,\n",
    "model.metabolites.cPY00129_c: -0.0030243,\n",
    "model.metabolites.lipid_c: 1})\n",
    "model.add_reactions([lipid_rxn])\n",
    "\n",
    "#Create biomass metabolite\n",
    "cpd11416_c = Metabolite('cpd11416_c',\n",
    "                        formula = '',\n",
    "                        name = 'Biomass',\n",
    "                        compartment = 'cytosol')\n",
    "model.add_metabolites([cpd11416_c])\n",
    "model.metabolites.cpd11416_c.annotation = {'seed.compound':'cpd11416','sbo': 'SBO:0000247'}\n",
    "model.metabolites.cpd11416_c.notes = {'NOTES':'*DDP 11/17/20 - added to represent biomass accumulation'}\n",
    "\n",
    "#Create biomass sink\n",
    "SK_cpd11416_c = Reaction('SK_cpd11416_c')\n",
    "SK_cpd11416_c.name = 'Biomass sink'\n",
    "SK_cpd11416_c.subsystem = 'cytosol'\n",
    "SK_cpd11416_c.lower_bound = 0\n",
    "SK_cpd11416_c.upper_bound = 1000\n",
    "SK_cpd11416_c.annotation = {'sbo': 'SBO:0000632'}\n",
    "SK_cpd11416_c.add_metabolites({\n",
    "model.metabolites.cpd11416_c: -1,})\n",
    "model.add_reactions([SK_cpd11416_c])\n",
    "\n",
    "#BOF: Remove metabolites corresponding to the binned categories and add the binned metabolites\n",
    "model.reactions.PA14_Biomass.add_metabolites({\n",
    "model.metabolites.cpd00084_c: 0.037574, #remove protein metabolites\n",
    "model.metabolites.cpd00065_c: 0.055748,\n",
    "model.metabolites.cpd00060_c: 0.076901,\n",
    "model.metabolites.cpd00119_c: 0.081188,\n",
    "model.metabolites.cpd00069_c: 0.094418,\n",
    "model.metabolites.cpd00132_c: 0.098829,\n",
    "model.metabolites.cpd00039_c: 0.10776,\n",
    "model.metabolites.cpd00066_c: 0.13212,\n",
    "model.metabolites.cpd00322_c: 0.15583,\n",
    "model.metabolites.cpd00161_c: 0.15734,\n",
    "model.metabolites.cpd00053_c: 0.16033,\n",
    "model.metabolites.cpd00129_c: 0.18965,\n",
    "model.metabolites.cpd00041_c: 0.19864,\n",
    "model.metabolites.cpd00054_c: 0.20723,\n",
    "model.metabolites.cpd00023_c: 0.22687,\n",
    "model.metabolites.cpd00156_c: 0.25557,\n",
    "model.metabolites.cpd00051_c: 0.28443,\n",
    "model.metabolites.cpd00033_c: 0.31256,\n",
    "model.metabolites.cpd00035_c: 0.43461,\n",
    "model.metabolites.cpd00107_c: 0.46321,\n",
    "model.metabolites.protein_c: -1, #replace with binned protein metabolite\n",
    "model.metabolites.cpd00294_c: 0.015972, #remove DNA metabolites\n",
    "model.metabolites.cpd00298_c: 0.015972,\n",
    "model.metabolites.cpd00206_c: 0.031422,\n",
    "model.metabolites.cpd00296_c: 0.031422,\n",
    "model.metabolites.dna_c: -1,    #replace with binned DNA metabolite\n",
    "model.metabolites.cpd00091_c: 0.10137,#remove RNA metabolites\n",
    "model.metabolites.cpd00018_c: 0.10146,\n",
    "model.metabolites.cpd00126_c: 0.19942,\n",
    "model.metabolites.cpd00046_c: 0.19949,\n",
    "model.metabolites.rna_c: -1,    #replace with binned RNA metabolite\n",
    "model.metabolites.cpd15430_c: 0.00037815, #remove lipid metabolites\n",
    "model.metabolites.cpd15540_c: 0.00042166,\n",
    "model.metabolites.cpd15533_c: 0.0020692,\n",
    "model.metabolites.cpd15539_c: 0.0037982,\n",
    "model.metabolites.cpd15429_c: 0.0041649,\n",
    "model.metabolites.cpd15428_c: 0.0078591,\n",
    "model.metabolites.cpd15541_c: 0.011557,\n",
    "model.metabolites.cpd15431_c: 0.011889,\n",
    "model.metabolites.cpd15538_c: 0.012557,\n",
    "model.metabolites.cpd15532_c: 0.020287,\n",
    "model.metabolites.cpd17066_c: 0.025523,\n",
    "model.metabolites.cpd15531_c: 0.047824,\n",
    "model.metabolites.cpd15534_c: 0.068198,\n",
    "model.metabolites.cPY00138_c: 0.00053681,\n",
    "model.metabolites.cPY00132_c: 0.00079222,\n",
    "model.metabolites.cPY00140_c: 0.00097843,\n",
    "model.metabolites.cPY00124_c: 0.0013602,\n",
    "model.metabolites.cPY00135_c: 0.0015844,\n",
    "model.metabolites.cPY00129_c: 0.0030243,\n",
    "model.metabolites.lipid_c: -1, #replace with binned lipid metabolite\n",
    "model.metabolites.cpd11416_c: 1, #add biomass metabolites\n",
    "model.metabolites.cpd17065_c: -0.001 #add LPS to OF\n",
    "})\n",
    "\n",
    "#Update model metadata\n",
    "model.id = 'iPau21'\n",
    "model.name = 'Pseudomonas aeruginosa UCBPP-PA14'\n",
    "model.annotation = {'taxonomy': '208963'}\n",
    "model.notes = {'creators':'Jennifer Bartell (iPau1129), Anna Blazier (iPau1129), Phillip Yen (iPau1129), Dawson Payne, Alina Renz, Laura Dunphy, Taylor Lewis, Jason Papin'}\n",
    "\n",
    "#Set media to blank\n",
    "for i in model.exchanges:\n",
    "    i.lower_bound = 0\n",
    "    i.upper_bound = 1000\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
